<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <!-- jQuery -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.10.3/jquery-ui.min.js"></script>
  
  <!-- bootstrap -->
  <!--<link href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" rel="stylesheet"  id="style">-->
  <script src="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/js/bootstrap.min.js"></script>
  
  <!-- highlight.js -->
  <!--<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" rel="stylesheet" id="code-style">-->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/highlight.min.js"></script>
  <script>
  hljs.LANGUAGES.r=function(a){var b="([a-zA-Z]|\\.[a-zA-Z.])[a-zA-Z0-9._]*";return{c:[a.HCM,{b:b,l:b,k:{keyword:"function if in break next repeat else for return switch while try tryCatch|10 stop warning require library attach detach source setMethod setGeneric setGroupGeneric setClass ...|10",literal:"NULL NA TRUE FALSE T F Inf NaN NA_integer_|10 NA_real_|10 NA_character_|10 NA_complex_|10"},r:0},{cN:"number",b:"0[xX][0-9a-fA-F]+[Li]?\\b",r:0},{cN:"number",b:"\\d+(?:[eE][+\\-]?\\d*)?L\\b",r:0},{cN:"number",b:"\\d+\\.(?!\\d)(?:i\\b)?",r:0},{cN:"number",b:"\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[a.BE],r:0},{cN:"string",b:"'",e:"'",c:[a.BE],r:0}]}}(hljs); </script>
  <!--<script type="text/javascript", src="https://yandex.st/highlightjs/7.3/languages/r.min.js"></script>-->
  
  <!-- Manific Popup -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.8.9/jquery.magnific-popup.min.js"></script>
  
  <script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script>
  <script defer="defer">
  // Function to generate the dynamic table of contents
  jQuery.fn.generate_TOC = function () {
    var base = $(this[0]);
  
    var selectors = ['h1', 'h2', 'h3', 'h4'];
  
    var last_ptr = [{}, {}, {}, {}];
  
    var anchors = {};
  
    generate_anchor = function (text) {
      var test = text.replace(/\W/g, '_');
  
      while(test in anchors){
        //if no suffix, add one
        if(test.match(/_\d+$/) === null){
          test = test + "_2";
        }
        //else generate unique id for duplicates by adding one to the suffix
        else {
          test = test.replace(/_(\d+)$/, function(match, number){ var num=+number+1; return("_" + num) });
        }
      }
      anchors[test]=1;
      return(test);
    }
  
    $(selectors.join(',')).filter(function(index) { return $(this).parent().attr("id") != 'header'; }).each(function () {
  
      var heading = $(this);
      var idx = selectors.indexOf(heading.prop('tagName').toLowerCase());
      var itr = 0;
  
      while (itr <= idx) {
        if (jQuery.isEmptyObject(last_ptr[itr])) {
          last_ptr[itr] = $('<ul>').addClass('nav');
          if (itr === 0) {
            base.append(last_ptr[itr])
          } else {
            if(last_ptr[itr-1].children('li').length === 0){
              last_ptr[itr-1].append(last_ptr[itr]);
            }
            else {
              last_ptr[itr - 1].children('li').last().append(last_ptr[itr]);
            }
          }
        }
        itr++;
      }
      var anchor = generate_anchor(heading.text());
      heading.attr('id', anchor);
      var a = $('<a>')
      .text(heading.text())
      .attr('href', '#' + anchor);
  
    var li = $('<li>')
      .append(a);
  
    last_ptr[idx].append(li);
    for (i = idx + 1; i < last_ptr.length; i++) {
      last_ptr[i] = {};
    }
    });
  }
  /* run scripts when document is ready */
  $(function() {
    "use strict";
  
    var $window = $(window);
    var $body = $(document.body);
  
    /* size of thumbnails */
  
    var hidden_types = ['source']
    var output_types = ['output', 'message', 'warning', 'error']
  
    /* style tables */
    $('table').addClass('table table-striped table-bordered table-hover table-condensed');
  
    $('pre code').each(function(i, e) {
      hljs.highlightBlock(e);
    });
  
    /* Magnific Popup */
    $(".thumbnail").each(function(){
      $(this).magnificPopup({
        disableOn: 768,
        closeOnContentClick: true,
  
        type: 'image',
        items: {
          src: $(this).find('img').attr('src'),
        }
      });
    });
  
    function toggle_block(obj, show) {
      var span = obj.find('span');
      if(show === true){
        span.removeClass('glyphicon-chevron-up').addClass('glyphicon-chevron-down');
        obj.next('pre').slideDown();
      }
      else {
        span.removeClass('glyphicon-chevron-down').addClass('glyphicon-chevron-up');
        obj.next('pre').slideUp();
      }
    }
  
    function toggle_thumbnails(imgs, show){
      if(show === true){
        imgs.parents().show()
        imgs.slideDown();
      }
      else {
        imgs.slideUp(400, function(){ $(this).parent().hide(); });
      }
    }
  
    function global_toggle(obj){
      var type = obj.attr('type');
      var show = !obj.parent('li').hasClass('active');
      if(show === true){
        obj.parent('li').addClass('active');
      }
      else{
        obj.parent('li').removeClass('active');
      }
      if(type == 'figure'){
        toggle_thumbnails($('.thumbnail img'), show);
      }
      else {
        $('.toggle.' + type).each(function() { toggle_block($(this), show); });
      }
    }
  
    /* onclick toggle next code block */
    $('.toggle').click(function() {
      var span = $(this).find('span');
      toggle_block($(this), !span.hasClass('glyphicon-chevron-down'));
      return false
    })
  
    // global toggles
    $('.toggle-global').click(function(){
      var type = $(this).attr('type');
      if(type === 'all-source'){
          $('li a.source').each(function() {
            global_toggle($(this));
          });
        }
      else if(type === 'all-output'){
        $.each(output_types, function(i, val){
          console.log(val);
          global_toggle($('li a.' + val));
        });
      }
      else {
        console.log($(this));
        global_toggle($(this));
      }
      return false;
    });
    /* table of contents */
    if($(['h1', 'h2', 'h3', 'h4'].join(',')).length > 0){
      $('body > #wrap > .container > .row').append('<div class="col-md-2"><div id="toc" class="well sidebar sidenav affix hidden-print"/></div>');
      $('#toc').generate_TOC();
    }
  
    $.each(hidden_types, function(i, type) {
      $('li[type=' + type + ']').each(function(){ global_toggle($(this)); });
    });
  
    /* remove paragraphs with no content */
    $('p:empty').remove();
  
    $body.scrollspy({
      target: '.sidebar',
    });
  
    /* theme switch */
    $('.theme-switch').click(function(){
      var css = $('link[title=' + $(this).attr('title') + ']');
      $('#theme[rel=stylesheet]').attr('href', css.attr('href'));
      $('.theme-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
    /* code style switch */ //TODO use same function for both of these?
    $('.highlight-switch').click(function(){
      var css = $('link[title="' + $(this).attr('title') + '"]');
      $('#highlight[rel=stylesheet]').attr('href', css.attr('href'));
      $('.highlight-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
  
    //TODO refresh on show/hide
    $window.on('load', function () {
      $body.scrollspy('refresh');
    })
  
  });
  
  </script>
  <style>
  /* Knitr_bootstrap styles */
  #header {
    display: none !important;
    visibility: hidden !important;
  }
  #wrap .container-fluid {
    padding: 0;
    overflow: hidden;
  }
  .toggle{
    text-transform: capitalize;
  }
  
  .toggle-global{
    text-transform: capitalize;
  }
  
  /* Sticky footer styles */
  * {
    margin:0;
  }
  html,
  body {
      height: 100%;
      padding:0 !important;
      /* The html and body elements cannot have any padding or margin. */
      /*overflow-x: hidden;*/
  }
  
  /* Wrapper for page content to push down footer */
  #wrap {
      min-height: 100%;
      height: auto !important;
      height: 100%;
      /* Negative indent footer by it's height */
      margin: 0 auto -120px;
  }
  
  /* Set the fixed height of the footer here */
  #push,
  #footer {
      height: 120px;
  }
  
  #footer {
    text-align: center;
  }
  
  /* Top level subheader elements.  These are the first nested items underneath a header element. */
  .header li {
    font-size: 20px;
  }
  
  /* Makes the font smaller for all subheader elements. */
  .sub-header li {
      font-size: 12px;
  }
  
  button.thumbnails {
    margin-left:0px;
  }
  
  /*
   * Side navigation
   *
   * Scrollspy and affixed enhanced navigation to highlight sections and secondary
   * sections of docs content.
   */
  
  /* By default it's not affixed in mobile views, so undo that */
  .sidebar.affix {
    position: static;
  }
  
  /* First level of nav */
  .sidenav {
    margin-top: 30px;
    margin-bottom: 30px;
    padding-top:    10px;
    padding-bottom: 10px;
    border-radius: 5px;
  }
  
  /* All levels of nav */
  .sidebar .nav > li > a {
    display: block;
    padding: 5px 20px;
  }
  .sidebar .nav > li > a:hover,
  .sidebar .nav > li > a:focus {
    text-decoration: none;
    border-right: 1px solid;
  }
  .sidebar .nav > .active > a,
  .sidebar .nav > .active:hover > a,
  .sidebar .nav > .active:focus > a {
    font-weight: bold;
    background-color: transparent;
    border-right: 1px solid;
  }
  
  /* Nav: second level (shown on .active) */
  .sidebar .nav .nav {
    display: none; /* Hide by default, but at >768px, show it */
    margin-bottom: 8px;
  }
  .sidebar .nav .nav > li > a {
    padding-top:    3px;
    padding-bottom: 3px;
    padding-left: 30px;
    font-size: 90%;
  }
  
  .sidebar .nav .nav .nav > li > a {
    padding-left: 40px;
  }
  .sidebar .nav .nav .nav .nav > li > a {
    padding-left: 50px;
  }
  
  /* Show and affix the side nav when space allows it */
  @media screen and (min-width: 992px) {
    .sidebar .nav > .active > ul {
      display: block;
    }
    /* Widen the fixed sidebar */
    .sidebar.affix,
    .sidebar.affix-bottom {
      width: 213px;
    }
    .sidebar.affix-top,
    .sidebar.affix {
      position: fixed; /* Undo the static from mobile first approach */
      top: 30px;
    }
    .sidebar.affix-bottom {
      position: absolute; /* Undo the static from mobile first approach */
    }
    .sidebar.affix-bottom .sidenav,
    .sidebar.affix .sidenav {
      margin-top: 0;
      margin-bottom: 0;
    }
  }
  @media screen and (min-width: 1200px) {
    /* Widen the fixed sidebar again */
    .sidebar.affix-bottom,
    .sidebar.affix {
      width: 263px;
    }
  }
  
  #toc {
    padding: 10px 0px;
    margin:0;
    border:0;
  }
  
  
  .panel pre {
    margin: 0;
    padding: 0;
    border: 0;
  }
  button + pre {
    margin: 0;
    padding: 0;
  }
  pre code {
    border-radius: 0;
  }
  /* Magnific Popup CSS */
  .mfp-bg {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1042;
    overflow: hidden;
    position: fixed;
    background: #0b0b0b;
    opacity: 0.8;
    filter: alpha(opacity=80); }
  
  .mfp-wrap {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1043;
    position: fixed;
    outline: none !important;
    -webkit-backface-visibility: hidden; }
  
  .mfp-container {
    text-align: center;
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    padding: 0 8px;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box; }
  
  .mfp-container:before {
    content: '';
    display: inline-block;
    height: 100%;
    vertical-align: middle; }
  
  .mfp-align-top .mfp-container:before {
    display: none; }
  
  .mfp-content {
    position: relative;
    display: inline-block;
    vertical-align: middle;
    margin: 0 auto;
    text-align: left;
    z-index: 1045; }
  
  .mfp-inline-holder .mfp-content,
  .mfp-ajax-holder .mfp-content {
    width: 100%;
    cursor: auto; }
  
  .mfp-ajax-cur {
    cursor: progress; }
  
  .mfp-zoom-out-cur,
  .mfp-zoom-out-cur .mfp-image-holder .mfp-close {
    cursor: -moz-zoom-out;
    cursor: -webkit-zoom-out;
    cursor: zoom-out; }
  
  .mfp-zoom {
    cursor: pointer;
    cursor: -webkit-zoom-in;
    cursor: -moz-zoom-in;
    cursor: zoom-in; }
  
  .mfp-auto-cursor .mfp-content {
    cursor: auto; }
  
  .mfp-close,
  .mfp-arrow,
  .mfp-preloader,
  .mfp-counter {
    -webkit-user-select: none;
    -moz-user-select: none;
    user-select: none; }
  
  .mfp-loading.mfp-figure {
    display: none; }
  
  .mfp-hide {
    display: none !important; }
  
  .mfp-preloader {
    color: #cccccc;
    position: absolute;
    top: 50%;
    width: auto;
    text-align: center;
    margin-top: -0.8em;
    left: 8px;
    right: 8px;
    z-index: 1044; }
  
  .mfp-preloader a {
    color: #cccccc; }
  
  .mfp-preloader a:hover {
    color: white; }
  
  .mfp-s-ready .mfp-preloader {
    display: none; }
  
  .mfp-s-error .mfp-content {
    display: none; }
  
  button.mfp-close,
  button.mfp-arrow {
    overflow: visible;
    cursor: pointer;
    background: transparent;
    border: 0;
    -webkit-appearance: none;
    display: block;
    padding: 0;
    z-index: 1046;
    -webkit-box-shadow: none;
    box-shadow: none; }
  
  button::-moz-focus-inner {
    padding: 0;
    border: 0; }
  
  .mfp-close {
    width: 44px;
    height: 44px;
    line-height: 44px;
    position: absolute;
    right: 0;
    top: 0;
    text-decoration: none;
    text-align: center;
    opacity: 0.65;
    padding: 0 0 18px 10px;
    color: white;
    font-style: normal;
    font-size: 28px;
    font-family: Arial, Baskerville, monospace; }
    .mfp-close:hover, .mfp-close:focus {
      opacity: 1; }
    .mfp-close:active {
      top: 1px; }
  
  .mfp-close-btn-in .mfp-close {
    color: #333333; }
  
  .mfp-image-holder .mfp-close,
  .mfp-iframe-holder .mfp-close {
    color: white;
    right: -6px;
    text-align: right;
    padding-right: 6px;
    width: 100%; }
  
  .mfp-counter {
    position: absolute;
    top: 0;
    right: 0;
    color: #cccccc;
    font-size: 12px;
    line-height: 18px; }
  
  .mfp-arrow {
    position: absolute;
    opacity: 0.65;
    margin: 0;
    top: 50%;
    margin-top: -55px;
    padding: 0;
    width: 90px;
    height: 110px;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); }
  
  .mfp-arrow:active {
    margin-top: -54px; }
  
  .mfp-arrow:hover,
  .mfp-arrow:focus {
    opacity: 1; }
  
  .mfp-arrow:before, .mfp-arrow:after,
  .mfp-arrow .mfp-b,
  .mfp-arrow .mfp-a {
    content: '';
    display: block;
    width: 0;
    height: 0;
    position: absolute;
    left: 0;
    top: 0;
    margin-top: 35px;
    margin-left: 35px;
    border: medium inset transparent; }
  .mfp-arrow:after,
  .mfp-arrow .mfp-a {
    border-top-width: 13px;
    border-bottom-width: 13px;
    top: 8px; }
  .mfp-arrow:before,
  .mfp-arrow .mfp-b {
    border-top-width: 21px;
    border-bottom-width: 21px; }
  
  .mfp-arrow-left {
    left: 0; }
    .mfp-arrow-left:after,
    .mfp-arrow-left .mfp-a {
      border-right: 17px solid white;
      margin-left: 31px; }
    .mfp-arrow-left:before,
    .mfp-arrow-left .mfp-b {
      margin-left: 25px;
      border-right: 27px solid #3f3f3f; }
  
  .mfp-arrow-right {
    right: 0; }
    .mfp-arrow-right:after,
    .mfp-arrow-right .mfp-a {
      border-left: 17px solid white;
      margin-left: 39px; }
    .mfp-arrow-right:before,
    .mfp-arrow-right .mfp-b {
      border-left: 27px solid #3f3f3f; }
  
  .mfp-iframe-holder {
    padding-top: 40px;
    padding-bottom: 40px; }
  
  .mfp-iframe-holder .mfp-content {
    line-height: 0;
    width: 100%;
    max-width: 900px; }
  
  .mfp-iframe-scaler {
    width: 100%;
    height: 0;
    overflow: hidden;
    padding-top: 56.25%; }
  
  .mfp-iframe-scaler iframe {
    position: absolute;
    display: block;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: black; }
  
  .mfp-iframe-holder .mfp-close {
    top: -40px; }
  
  /* Main image in popup */
  img.mfp-img {
    width: auto;
    max-width: 100%;
    height: auto;
    display: block;
    line-height: 0;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    padding: 40px 0 40px;
    margin: 0 auto; }
  
  /* The shadow behind the image */
  .mfp-figure:after {
    content: '';
    position: absolute;
    left: 0;
    top: 40px;
    bottom: 40px;
    display: block;
    right: 0;
    width: auto;
    height: auto;
    z-index: -1;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: #444444; }
  
  .mfp-figure {
    line-height: 0; }
  
  .mfp-bottom-bar {
    margin-top: -36px;
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    cursor: auto; }
  
  .mfp-title {
    text-align: left;
    line-height: 18px;
    color: #f3f3f3;
    word-wrap: break-word;
    padding-right: 36px; }
  
  .mfp-figure small {
    color: #bdbdbd;
    display: block;
    font-size: 12px;
    line-height: 14px; }
  
  .mfp-image-holder .mfp-content {
    max-width: 100%; }
  
  .mfp-gallery .mfp-image-holder .mfp-figure {
    cursor: pointer; }
  
  @media screen and (max-width: 800px) and (orientation: landscape), screen and (max-height: 300px) {
    /**
     * Remove all paddings around the image on small screen
     */
    .mfp-img-mobile .mfp-image-holder {
      padding-left: 0;
      padding-right: 0; }
  
    .mfp-img-mobile img.mfp-img {
      padding: 0; }
  
    /* The shadow behind the image */
    .mfp-img-mobile .mfp-figure:after {
      top: 0;
      bottom: 0; }
  
    .mfp-img-mobile .mfp-bottom-bar {
      background: rgba(0, 0, 0, 0.6);
      bottom: 0;
      margin: 0;
      top: auto;
      padding: 3px 5px;
      position: fixed;
      -webkit-box-sizing: border-box;
      -moz-box-sizing: border-box;
      box-sizing: border-box; }
  
    .mfp-img-mobile .mfp-bottom-bar:empty {
      padding: 0; }
  
    .mfp-img-mobile .mfp-counter {
      right: 5px;
      top: 3px; }
  
    .mfp-img-mobile .mfp-close {
      top: 0;
      right: 0;
      width: 35px;
      height: 35px;
      line-height: 35px;
      background: rgba(0, 0, 0, 0.6);
      position: fixed;
      text-align: center;
      padding: 0; }
  
    .mfp-img-mobile .mfp-figure small {
      display: inline;
      margin-left: 5px; } }
  @media all and (max-width: 900px) {
    .mfp-arrow {
      -webkit-transform: scale(0.75);
      transform: scale(0.75); }
  
    .mfp-arrow-left {
      -webkit-transform-origin: 0;
      transform-origin: 0; }
  
    .mfp-arrow-right {
      -webkit-transform-origin: 100%;
      transform-origin: 100%; }
  
    .mfp-container {
      padding-left: 6px;
      padding-right: 6px; } }
  .mfp-ie7 .mfp-img {
    padding: 0; }
  .mfp-ie7 .mfp-bottom-bar {
    width: 600px;
    left: 50%;
    margin-left: -300px;
    margin-top: 5px;
    padding-bottom: 5px; }
  .mfp-ie7 .mfp-container {
    padding: 0; }
  .mfp-ie7 .mfp-content {
    padding-top: 44px; }
  .mfp-ie7 .mfp-close {
    top: 0;
    right: 0;
    padding-top: 0; }
  
  //Magnific overrides
  .mfp-image img{
    background: white;
  }
  .mfp-figure:after {
    background: white;
  }
  
  /*
   * Off Canvas navbar toggle right
   * --------------------------------------------------
   */
  
  @media screen and (max-width: 768px) {
    .row-offcanvas .collapsing {
    -webkit-transition: none 0;
      -moz-transition: none 0;
      transition: none 0;
    }
   .row-offcanvas .navbar {
    position: absolute;
    z-index: 2;
      right:0;
      height:100%;
      width:55px;
      border:0;
      background-color:transparent;
    }
    .row-offcanvas .navbar-toggle {
      margin-right: 5px;
      margin-left: 5px;
    }
    .row-offcanvas {
      position: relative;
    }
    .row-offcanvas-right.active .navbar {
    position: absolute;
    z-index: 2;
      right: -28.4%;
      width:40%;
      background-color:#eee;
      border:0 solid #ddd;
      border-left-width:1px;
    }
    .row-offcanvas-right.active {
      left: -30%;
    }
    .row-offcanvas-right.active .navbar-collapse {
      position: relative;
      width: 100%;
    }
    .row-offcanvas .content {
    /*width:calc(100% - 60px);*/
    }
  }
  </style>
</head>
<body>
<div id="wrap">
<div class="container">
<div class="row row-offcanvas row-offcanvas-right">
<div class="contents col-xs-12 col-md-10">
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Introduction to derfinder}
-->
<h1>Introduction to <code>derfinder</code></h1>
<p>If you wish, you can view this vignette online <a href="http://lcolladotor.github.io/derfinder/">here</a>.</p>
<h1>Overview</h1>
<p>In this vignette we present <code>derfinder</code> <a id='cite-derfinder'></a>(<a href='https://github.com/lcolladotor/derfinder'>Collado-Torres, Frazee, Jaffe, and Leek, 2014</a>) which enables annotation-agnostic differential expression analysis at base-pair resolution. A example with publicly available data is showcased in this vignette. With it, the two main type of analyses that can be performed with <code>derfinder</code> are illustrated.</p>
<h1>Example</h1>
<p>As an example, we will analyze a small subset of the samples from the <em>BrainSpan Atlas of the Human Brain</em> <a id='cite-brainspan'></a>(<a href='http://developinghumanbrain.org'>BrainSpan, 2011</a>) publicly available data.</p>
<p>We first load the required packages.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Load libraries
library('derfinder')
library('derfinderData')</code></pre>
</div>
<h2>Phenotype data</h2>
<p>For this example, we created a small table with the relevant phenotype data for 12 samples: 6 from fetal samples and 6 from adult samples. We chose at random a brain region, in this case the amygdaloid complex. For this example we will only look at data from chromosome 21. Other variables include the age in years, the gender and the RNA Integrity Number (RIN). The data is shown below.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-up"></span> R source
</button>
<pre style="display:none"><code class="source r">library('knitr')
## Get pheno table
pheno <- subset(brainspanPheno, structure_acronym == 'AMY')

## Display the main information
p <- pheno[, -which(colnames(pheno) %in% c('structure_acronym', 'structure_name', 'file'))]
rownames(p) <- NULL
kable(p, format = 'html', row.names = TRUE)</code></pre>
<table>
<thead>
<tr>
<th align="left">
</th>
<th align="left">
gender
</th>
<th align="left">
lab
</th>
<th align="right">
Age
</th>
<th align="right">
RIN
</th>
<th align="left">
group
</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">
1
</td>
<td align="left">
F
</td>
<td align="left">
HSB97.AMY
</td>
<td align="right">
-0.5476
</td>
<td align="right">
9.1
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
2
</td>
<td align="left">
M
</td>
<td align="left">
HSB92.AMY
</td>
<td align="right">
-0.4524
</td>
<td align="right">
9.2
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
3
</td>
<td align="left">
M
</td>
<td align="left">
HSB178.AMY
</td>
<td align="right">
-0.5714
</td>
<td align="right">
9.8
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
4
</td>
<td align="left">
M
</td>
<td align="left">
HSB159.AMY
</td>
<td align="right">
-0.3810
</td>
<td align="right">
9.9
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
5
</td>
<td align="left">
F
</td>
<td align="left">
HSB153.AMY
</td>
<td align="right">
-0.6667
</td>
<td align="right">
9.3
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
6
</td>
<td align="left">
F
</td>
<td align="left">
HSB113.AMY
</td>
<td align="right">
-0.6667
</td>
<td align="right">
9.4
</td>
<td align="left">
fetal
</td>
</tr>
<tr>
<td align="left">
7
</td>
<td align="left">
F
</td>
<td align="left">
HSB130.AMY
</td>
<td align="right">
21.0000
</td>
<td align="right">
8.6
</td>
<td align="left">
adult
</td>
</tr>
<tr>
<td align="left">
8
</td>
<td align="left">
M
</td>
<td align="left">
HSB136.AMY
</td>
<td align="right">
23.0000
</td>
<td align="right">
8.1
</td>
<td align="left">
adult
</td>
</tr>
<tr>
<td align="left">
9
</td>
<td align="left">
F
</td>
<td align="left">
HSB126.AMY
</td>
<td align="right">
30.0000
</td>
<td align="right">
8.4
</td>
<td align="left">
adult
</td>
</tr>
<tr>
<td align="left">
10
</td>
<td align="left">
M
</td>
<td align="left">
HSB145.AMY
</td>
<td align="right">
36.0000
</td>
<td align="right">
7.4
</td>
<td align="left">
adult
</td>
</tr>
<tr>
<td align="left">
11
</td>
<td align="left">
M
</td>
<td align="left">
HSB123.AMY
</td>
<td align="right">
37.0000
</td>
<td align="right">
7.5
</td>
<td align="left">
adult
</td>
</tr>
<tr>
<td align="left">
12
</td>
<td align="left">
F
</td>
<td align="left">
HSB135.AMY
</td>
<td align="right">
40.0000
</td>
<td align="right">
8.5
</td>
<td align="left">
adult
</td>
</tr>
</tbody>
</table>
</div>
<h2>Load the data</h2>
<p><code>derfinder</code> offers three functions related to loading data. The first one, <code>rawFiles()</code>, is a helper function for identifying the full paths to the input files. Next, <code>loadCoverage()</code> loads the base-level coverage data from either BAM or BigWig files for a specific chromosome. Finally, <code>fullCoverage()</code> will load the coverage for a set of chromosomes using <code>loadCoverage()</code>.</p>
<p>We can load the data from <code>derfinderData</code> <a id='cite-derfinderData'></a>(<a href='https://github.com/lcolladotor/derfinderData'>Collado-Torres, Jaffe, and Leek, 2014</a>) by first identifying the paths to the BigWig files with <code>rawFiles()</code> and then loading the data with <code>fullCoverage()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Determine the files to use and fix the names
files <- rawFiles(system.file('extdata', 'AMY', package = 'derfinderData'), samplepatt = 'bw', fileterm = NULL)
names(files) <- gsub('.bw', '', names(files))

## Load the data from disk
system.time(fullCov <- fullCoverage(files = files, chrs = 'chr21'))</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:45:57 fullCoverage: processing chromosome chr21
## 2014-09-30 14:45:57 loadCoverage: finding chromosome lengths
## 2014-09-30 14:45:57 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB113.bw
## Loading required package: parallel
## 
## Attaching package: 'BiocGenerics'
## 
## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
## 
## The following object is masked from 'package:stats':
## 
##     xtabs
## 
## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, as.vector, cbind,
##     colnames, do.call, duplicated, eval, evalq, Filter, Find, get,
##     intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rep.int, rownames, sapply, setdiff, sort,
##     table, tapply, union, unique, unlist, unsplit
## 
## Loading required package: stats4
## 2014-09-30 14:45:57 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB123.bw
## 2014-09-30 14:45:58 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB126.bw
## 2014-09-30 14:45:58 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB130.bw
## 2014-09-30 14:45:58 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB135.bw
## 2014-09-30 14:45:59 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB136.bw
## 2014-09-30 14:45:59 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB145.bw
## 2014-09-30 14:45:59 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB153.bw
## 2014-09-30 14:46:00 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB159.bw
## 2014-09-30 14:46:00 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB178.bw
## 2014-09-30 14:46:01 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB92.bw
## 2014-09-30 14:46:01 loadCoverage: loading BigWig file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinderData/extdata/AMY/HSB97.bw
## 2014-09-30 14:46:02 loadCoverage: applying the cutoff to the merged data
## 2014-09-30 14:46:02 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##   5.006   0.078   5.106
</code></pre>
</div>
<p>Alternatively, since the BigWig files are publicly available from <em>BrainSpan</em> (see <a href="http://download.alleninstitute.org/brainspan/MRF_BigWig_Gencode_v10/bigwig/">here</a>), we can extract the relevant coverage data using <code>fullCoverage()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Determine the files to use and fix the names
files <- pheno$file
names(files) <- gsub('.AMY', '', pheno$lab)

## Load the data from the web
system.time(fullCov <- fullCoverage(files = files, chrs = 'chr21'))</code></pre>
</div>
<p>Note how loading the coverage for 12 samples from the web was quite fast. Although in this case we only retained the information for chromosome 21.</p>
<p>The result of <code>fullCov</code> is a list with one element per chromosome. If no filtering was performed, each chromosome has a <code>DataFrame</code> with the number of rows equaling the number of bases in the chromosome with one column per sample.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Lets explore it
fullCov</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $chr21
## DataFrame with 48129895 rows and 12 columns
##          HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159
##           <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>
## 1             0      0      0      0      0      0      0      0      0
## 2             0      0      0      0      0      0      0      0      0
## 3             0      0      0      0      0      0      0      0      0
## 4             0      0      0      0      0      0      0      0      0
## 5             0      0      0      0      0      0      0      0      0
## ...         ...    ...    ...    ...    ...    ...    ...    ...    ...
## 48129891      0      0      0      0      0      0      0      0      0
## 48129892      0      0      0      0      0      0      0      0      0
## 48129893      0      0      0      0      0      0      0      0      0
## 48129894      0      0      0      0      0      0      0      0      0
## 48129895      0      0      0      0      0      0      0      0      0
##          HSB178 HSB92 HSB97
##           <Rle> <Rle> <Rle>
## 1             0     0     0
## 2             0     0     0
## 3             0     0     0
## 4             0     0     0
## 5             0     0     0
## ...         ...   ...   ...
## 48129891      0     0     0
## 48129892      0     0     0
## 48129893      0     0     0
## 48129894      0     0     0
## 48129895      0     0     0
</code></pre>
</div>
<p>If filtering was performed, each chromosome also has a logical <code>Rle</code> indicating which bases of the chromosome passed the filtering. This information is useful later on to map back the results to the genome coordinates.</p>
<h2>Filter coverage</h2>
<p>Depending on the use case, you might want to filter the base level coverage at the time of reading it, or you might want to keep an unfiltered version. By default both <code>loadCoverage()</code> and <code>fullCoverage()</code> will not filter.</p>
<p>If you decide to filter, set the <code>cutoff</code> argument to a positive value. This will run <code>filterData()</code>. Note that you might want to standardize the library sizes prior to filtering, which can be done by supplying the <code>totalMapped</code> and <code>targetSize</code> arguments.</p>
<p>In this example, we prefer to keep both an unfiltered and filtered version. For the filtered version, we will retain the bases where at least one sample has coverage greater than 2.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Filter coverage
filteredCov <- lapply(fullCov, filterData, cutoff = 2)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:03 filterData: originally there were 48129895 rows, now there are 130356 rows. Meaning that 99.73 percent was filtered.
</code></pre>
</div>
<p>The result is similar to <code>fullCov</code> but with the genomic position index as shown below.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Similar to fullCov but with $position
filteredCov</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $chr21
## $chr21$coverage
## DataFrame with 130356 rows and 12 columns
##        HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159
##         <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>  <Rle>
## 1        2.01      0   0.00   0.04   0.23   0.13      0   0.59   0.15
## 2        2.17      0   0.00   0.04   0.23   0.13      0   0.63   0.15
## 3        2.21      0   0.00   0.04   0.23   0.13      0   0.67   0.15
## 4        2.37      0   0.00   0.04   0.23   0.13      0   0.71   0.15
## 5        2.37      0   0.06   0.04   0.23   0.13      0   0.75   0.25
## ...       ...    ...    ...    ...    ...    ...    ...    ...    ...
## 130352   1.25   1.28   2.05   0.79   1.63   1.37   1.03   2.21   2.46
## 130353   1.21   1.24   2.05   0.79   1.63   1.37   1.03   2.21   2.46
## 130354   1.21   1.20   1.93   0.79   1.63   1.28   0.73   2.01   2.21
## 130355   1.17   1.20   1.93   0.79   1.63   1.28   0.60   2.01   2.11
## 130356   1.17   1.09   1.87   0.79   1.55   1.02   0.56   1.93   1.96
##        HSB178 HSB92 HSB97
##         <Rle> <Rle> <Rle>
## 1        0.05  0.07  1.58
## 2        0.05  0.07  1.58
## 3        0.05  0.10  1.61
## 4        0.05  0.13  1.65
## 5        0.10  0.13  1.68
## ...       ...   ...   ...
## 130352   2.07  2.23  1.51
## 130353   2.07  2.23  1.51
## 130354   2.11  2.13  1.47
## 130355   2.11  2.13  1.47
## 130356   1.78  2.06  1.47
## 
## $chr21$position
## logical-Rle of length 48129895 with 3235 runs
##   Lengths: 9825448     149       2       9 ...     740     598   45091
##   Values :   FALSE    TRUE   FALSE    TRUE ...   FALSE    TRUE   FALSE
</code></pre>
</div>
<p>In terms of memory, the filtered version requires less resources. Although this will depend on how rich the data set is and how aggressive was the filtering step.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Compare the size in Mb
round(c(fullCov = object.size(fullCov), filteredCov = object.size(filteredCov)) / 1024^2, 1)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##     fullCov filteredCov 
##        22.7         8.5
</code></pre>
</div>
<h2>DER analysis</h2>
<p>One form of base-level differential expression analysis implemented in <code>derfinder</code> is to calculate F-statistics for every base and use them to define candidate differentially expressed regions. This type of analysis is further explained in this section.</p>
<h3>Models</h3>
<p>Once we have the base-level coverage data for all 12 samples, we can construct the models. In this case, we want to find differences between fetal and adult samples while adjusting for gender, RIN and a proxy of the library size.</p>
<p>We can use <code>sampleDepth()</code> and it's helper function <code>collapseFullCoverage()</code> to get a proxy of the library size. Note that you would normally use the unfiltered data from all the chromosomes in this step and not just one.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Get some idea of the library sizes
sampleDepths <- sampleDepth(collapseFullCoverage(fullCov), 1)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:03 sampleDepth: Calculating sample quantiles
## 2014-09-30 14:46:09 sampleDepth: Calculating sample adjustments
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">sampleDepths</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## HSB113.100% HSB123.100% HSB126.100% HSB130.100% HSB135.100% HSB136.100% 
##       19.82       19.41       19.53       19.52       20.33       19.98 
## HSB145.100% HSB153.100% HSB159.100% HSB178.100%  HSB92.100%  HSB97.100% 
##       19.50       19.41       19.24       19.44       19.56       19.48
</code></pre>
</div>
<p><code>sampleDepth()</code> is similar to <code>calcNormFactors()</code> from <a href="http://www.bioconductor.org/packages/release/bioc/html/metagenomeSeq.html">metagenomeSeq</a> with some code underneath tailored for the type of data we are using. <code>collapseFullCoverage()</code> is only needed to deal with the size of the data.</p>
<p>We can then define the nested models we want to use using <code>makeModels()</code>. This is a helper function that assumes that you will always adjust for the library size. You then need to define the variable to test, in this case we are comparing fetal vs adult samples. Optionally, you can adjust for other sample covariates, such as the gender and RIN in this case.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Define models
models <- makeModels(sampleDepths, testvars = pheno$group, adjustvars = pheno[, c('gender', 'RIN')]) 

## Explore the models
lapply(models, head)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $mod
##   (Intercept) testvarsadult sampleDepths adjustVar1M adjustVar2
## 1           1             0        19.82           0        9.1
## 2           1             0        19.41           1        9.2
## 3           1             0        19.53           1        9.8
## 4           1             0        19.52           1        9.9
## 5           1             0        20.33           0        9.3
## 6           1             0        19.98           0        9.4
## 
## $mod0
##   (Intercept) sampleDepths adjustVar1M adjustVar2
## 1           1        19.82           0        9.1
## 2           1        19.41           1        9.2
## 3           1        19.53           1        9.8
## 4           1        19.52           1        9.9
## 5           1        20.33           0        9.3
## 6           1        19.98           0        9.4
</code></pre>
</div>
<p>Note how the null model (<code>mod0</code>) is nested in the alternative model (<code>mod</code>). <code>derfinder</code> is very flexible and works with any type of nested model.</p>
<h3>Find candidate DERs</h3>
<p>Next, we can find candidate differentially expressed regions (DERs) using as input the segments of the genome where at least one sample has coverage greater than 2. That is, the filtered coverage version we created previously.</p>
<p>The <strong>main</strong> function in <code>derfinder</code> for this type of analysis is <code>analyzeChr()</code>. It works at a chromosome level and runs behinds the scenes several other <code>derfinder</code> functions. To use it, you have to provide the models, the grouping information, how to calculate the F-statistic cutoff and most importantly, the number of permutations.</p>
<p>By default <code>analyzeChr()</code> will use a theoretical cutoff. In this example, we use the cutoff that would correspond to a p-value of 0.01. To assign p-values to the candidate DERs, <code>derfinder</code> permutes the rows of the model matrices, re-calculates the F-statistics and identifies null regions. Then it compares the area of the observed regions versus the areas from the null regions to assign an empirical p-value.</p>
<p>In this example we will use twenty permutations, although in a real case scenario you might consider using up to a thousand.</p>
<p>In real scenario, you might consider saving the results from all the chromosomes in a given directory. Here we will use <em>analysisResults</em>. For each chromosome you analyze, a new directory with the chromosome-specific data will be created. So in this case, we will have <em>analysisResults/chr21</em>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Create a analysis directory
dir.create('analysisResults')
originalWd <- getwd()
setwd(file.path(originalWd, 'analysisResults'))

## Perform differential expression analysis
system.time(results <- analyzeChr(chr = 'chr21', filteredCov$chr21, models, groupInfo = pheno$group, writeOutput = TRUE, cutoffFstat = 1e-02, nPermute = 20, seeds = 20140923 + seq_len(20), returnOutput = TRUE))</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:10 analyzeChr: Pre-processing the coverage data
## 2014-09-30 14:46:15 analyzeChr: Calculating statistics
## 2014-09-30 14:46:15 calculateStats: calculating the F-statistics
## 2014-09-30 14:46:16 analyzeChr: Calculating pvalues
## 2014-09-30 14:46:16 analyzeChr: Using the following theoretical cutoff for the F-statistics 12.2463833484351
## 2014-09-30 14:46:16 calculatePvalues: identifying data segments
## 2014-09-30 14:46:16 findRegions: segmenting F-stats information
## 2014-09-30 14:46:16 findRegions: identifying candidate DERs
## 2014-09-30 14:46:16 findRegions: identifying DER clusters
## 2014-09-30 14:46:16 calculatePvalues: calculating F-statistics for permutation 1 and seed 20140924
## 2014-09-30 14:46:17 findRegions: segmenting F-stats information
## 2014-09-30 14:46:17 findRegions: identifying candidate DERs
## 2014-09-30 14:46:17 calculatePvalues: calculating F-statistics for permutation 2 and seed 20140925
## 2014-09-30 14:46:17 findRegions: segmenting F-stats information
## 2014-09-30 14:46:17 findRegions: identifying candidate DERs
## 2014-09-30 14:46:17 calculatePvalues: calculating F-statistics for permutation 3 and seed 20140926
## 2014-09-30 14:46:18 findRegions: segmenting F-stats information
## 2014-09-30 14:46:18 findRegions: identifying candidate DERs
## 2014-09-30 14:46:18 calculatePvalues: calculating F-statistics for permutation 4 and seed 20140927
## 2014-09-30 14:46:18 findRegions: segmenting F-stats information
## 2014-09-30 14:46:18 findRegions: identifying candidate DERs
## 2014-09-30 14:46:18 calculatePvalues: calculating F-statistics for permutation 5 and seed 20140928
## 2014-09-30 14:46:18 findRegions: segmenting F-stats information
## 2014-09-30 14:46:18 findRegions: identifying candidate DERs
## 2014-09-30 14:46:18 calculatePvalues: calculating F-statistics for permutation 6 and seed 20140929
## 2014-09-30 14:46:19 findRegions: segmenting F-stats information
## 2014-09-30 14:46:19 findRegions: identifying candidate DERs
## 2014-09-30 14:46:19 calculatePvalues: calculating F-statistics for permutation 7 and seed 20140930
## 2014-09-30 14:46:19 findRegions: segmenting F-stats information
## 2014-09-30 14:46:19 findRegions: identifying candidate DERs
## 2014-09-30 14:46:19 calculatePvalues: calculating F-statistics for permutation 8 and seed 20140931
## 2014-09-30 14:46:19 findRegions: segmenting F-stats information
## 2014-09-30 14:46:19 findRegions: identifying candidate DERs
## 2014-09-30 14:46:20 calculatePvalues: calculating F-statistics for permutation 9 and seed 20140932
## 2014-09-30 14:46:20 findRegions: segmenting F-stats information
## 2014-09-30 14:46:20 findRegions: identifying candidate DERs
## 2014-09-30 14:46:20 calculatePvalues: calculating F-statistics for permutation 10 and seed 20140933
## 2014-09-30 14:46:20 findRegions: segmenting F-stats information
## 2014-09-30 14:46:20 findRegions: identifying candidate DERs
## 2014-09-30 14:46:20 calculatePvalues: calculating F-statistics for permutation 11 and seed 20140934
## 2014-09-30 14:46:21 findRegions: segmenting F-stats information
## 2014-09-30 14:46:21 findRegions: identifying candidate DERs
## 2014-09-30 14:46:21 calculatePvalues: calculating F-statistics for permutation 12 and seed 20140935
## 2014-09-30 14:46:21 findRegions: segmenting F-stats information
## 2014-09-30 14:46:21 findRegions: identifying candidate DERs
## 2014-09-30 14:46:21 calculatePvalues: calculating F-statistics for permutation 13 and seed 20140936
## 2014-09-30 14:46:21 findRegions: segmenting F-stats information
## 2014-09-30 14:46:21 findRegions: identifying candidate DERs
## 2014-09-30 14:46:21 calculatePvalues: calculating F-statistics for permutation 14 and seed 20140937
## 2014-09-30 14:46:22 findRegions: segmenting F-stats information
## 2014-09-30 14:46:22 findRegions: identifying candidate DERs
## 2014-09-30 14:46:22 calculatePvalues: calculating F-statistics for permutation 15 and seed 20140938
## 2014-09-30 14:46:22 findRegions: segmenting F-stats information
## 2014-09-30 14:46:22 findRegions: identifying candidate DERs
## 2014-09-30 14:46:22 calculatePvalues: calculating F-statistics for permutation 16 and seed 20140939
## 2014-09-30 14:46:22 findRegions: segmenting F-stats information
## 2014-09-30 14:46:22 findRegions: identifying candidate DERs
## 2014-09-30 14:46:22 calculatePvalues: calculating F-statistics for permutation 17 and seed 20140940
## 2014-09-30 14:46:23 findRegions: segmenting F-stats information
## 2014-09-30 14:46:23 findRegions: identifying candidate DERs
## 2014-09-30 14:46:23 calculatePvalues: calculating F-statistics for permutation 18 and seed 20140941
## 2014-09-30 14:46:23 findRegions: segmenting F-stats information
## 2014-09-30 14:46:23 findRegions: identifying candidate DERs
## 2014-09-30 14:46:23 calculatePvalues: calculating F-statistics for permutation 19 and seed 20140942
## 2014-09-30 14:46:23 findRegions: segmenting F-stats information
## 2014-09-30 14:46:23 findRegions: identifying candidate DERs
## 2014-09-30 14:46:23 calculatePvalues: calculating F-statistics for permutation 20 and seed 20140943
## 2014-09-30 14:46:23 findRegions: segmenting F-stats information
## 2014-09-30 14:46:24 findRegions: identifying candidate DERs
## 2014-09-30 14:46:24 calculatePvalues: calculating the p-values
## 2014-09-30 14:46:24 analyzeChr: Annotating regions
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## Matching regions to genes.
## nearestgene: loading bumphunter hg19 transcript database
## finding nearest transcripts...
## AnnotatingDone.
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##  14.180   1.432  15.533
</code></pre>
</div>
<p>To speed up <code>analyzeChr()</code>, you might need to use several cores via the <code>mc.cores</code> argument. If memory is limiting, you might want to use a smaller <code>chunksize</code> (default is 5 million). Note that if you use too many cores, you might hit the input/output ceiling of your data network and/or hard drives speed.</p>
<p>Before using a large number of permutations we recommend exploring how long each permutation cycle takes using a single permutation.</p>
<h3>Explore results</h3>
<p>When using <code>returnOutput = TRUE</code>, <code>analyzeChr()</code> will return a list with the results to explore interactively. However, by default it writes the results to disk (one .Rdata file per result).</p>
<p>The following code explores the results.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore
names(results)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "timeinfo"     "optionsStats" "coveragePrep" "fstats"      
## [5] "regions"      "annotation"
</code></pre>
</div>
<h4>optionStats</h4>
<p><code>optionStats</code> stores the main options used in the <code>analyzeChr()</code> call including the models used, the type of cutoff, number of permutations, seeds for the permutations. All this information can be useful to reproduce the analysis.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore optionsStats
names(results$optionsStats)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  [1] "models"          "cutoffPre"       "scalefac"       
##  [4] "chunksize"       "cutoffFstat"     "cutoffType"     
##  [7] "nPermute"        "seeds"           "groupInfo"      
## [10] "lowMemDir"       "analyzeCall"     "cutoffFstatUsed"
## [13] "returnOutput"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Call used
results$optionsStats$analyzeCall</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## analyzeChr(chr = "chr21", coverageInfo = filteredCov$chr21, models = models, 
##     cutoffFstat = 0.01, nPermute = 20, seeds = 20140923 + seq_len(20), 
##     groupInfo = pheno$group, writeOutput = TRUE, returnOutput = TRUE)
</code></pre>
</div>
<h4>coveragePrep</h4>
<p><code>coveragePrep</code> has the result from the <code>preprocessCoverage()</code> step. This includes the genomic position index, the mean coverage (after scaling and the log2 transformation) for all the samples, and the group mean coverages. By default, the chunks are written to disk in <code>optionsStats$lowMemDir</code> (chr21/chunksDir in this example) to help reduce the required memory resources. Otherwise it is stored in <code>coveragePrep$coverageProcessed</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore coveragePrep
names(results$coveragePrep)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "coverageProcessed" "mclapplyIndex"     "position"         
## [4] "meanCoverage"      "groupMeans"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Group means
results$coveragePrep$groupMeans</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $fetal
## numeric-Rle of length 130356 with 116452 runs
##   Lengths:                 1                 1 ...                 1
##   Values : 0.401666664828857 0.428333345800638 ...  1.24833332498868
## 
## $adult
## numeric-Rle of length 130356 with 119226 runs
##   Lengths:                 1                 1 ...                 1
##   Values : 0.406666670615474  0.41333334085842 ...   1.6266666551431
</code></pre>
</div>
<h4>fstats</h4>
<p>The F-statistics are then stored in <code>fstats</code>. These are calculated using <code>calculateStats()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore optionsStats
results$fstats</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## numeric-Rle of length 130356 with 126807 runs
##   Lengths:                   1                   1 ...                   1
##   Values :   0.498752703304995     0.5868035061967 ...  0.0104137056203738
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Note that the length matches the number of bases used
identical(length(results$fstats), sum(results$coveragePrep$position))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] TRUE
</code></pre>
</div>
<h4>regions</h4>
<p>The candidate DERs and summary results from the permutations is then stored in <code>regions</code>. This is the output from <code>calculatePvalues()</code> which uses several underneath other functions including <code>calculateStats()</code> and <code>findRegions()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore regions
names(results$regions)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "regions"         "nullStats"       "nullWidths"      "nullPermutation"
</code></pre>
</div>
<p>For the null regions, the summary information is composed of the mean F-statistic for the null regions (<code>regions$nullStats</code>), the width of the null regions (<code>regions$nullWidths</code>), and the permutation number under which they were identified (<code>regions$nullPermutation</code>).</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Permutation summary information
results$regions[2:4]</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $nullStats
## numeric-Rle of length 2272 with 2272 runs
##   Lengths:                1                1 ...                1
##   Values : 12.2477014397316 14.9676808486673 ... 16.2530220307103
## 
## $nullWidths
## integer-Rle of length 2272 with 1988 runs
##   Lengths:   1   1   1   1   1   1   1   1 ...   1   1   1   1   1   1   1
##   Values :   1  21   1   4   1   3   2   5 ...   1   2   5   2   1   4  10
## 
## $nullPermutation
## integer-Rle of length 2272 with 20 runs
##   Lengths:  20 872  50 138  40 180  94  88 ... 122  56 174  70  24  12  28
##   Values :   1   2   3   4   5   6   7   8 ...  14  15  16  17  18  19  20
</code></pre>
</div>
<p>The most important part of the output is the <code>GRanges</code> object with the candidate DERs shown below.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Candidate DERs
results$regions$regions</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## GRanges object with 21 ranges and 14 metadata columns:
##       seqnames               ranges strand   |     value      area
##          <Rle>            <IRanges>  <Rle>   | <numeric> <numeric>
##    up    chr21 [45751762, 45751800]      *   |     20.08     783.1
##    up    chr21 [35210152, 35210183]      *   |     22.62     723.8
##    up    chr21 [46190082, 46190103]      *   |     20.47     450.4
##    up    chr21 [38574010, 38574032]      *   |     15.74     362.0
##    up    chr21 [45564707, 45564721]      *   |     16.72     250.8
##   ...      ...                  ...    ... ...       ...       ...
##    up    chr21 [45751879, 45751879]      *   |     12.85     12.85
##    up    chr21 [34975723, 34975723]      *   |     12.75     12.75
##    up    chr21 [38574058, 38574058]      *   |     12.53     12.53
##    up    chr21 [45751849, 45751849]      *   |     12.44     12.44
##    up    chr21 [46190080, 46190080]      *   |     12.31     12.31
##       indexStart  indexEnd cluster clusterL meanCoverage meanfetal
##        <integer> <integer>   <Rle>    <Rle>    <numeric> <numeric>
##    up     108257    108295       6      125        1.509    1.5249
##    up      55566     55597       3      151        1.079    0.9185
##    up     109456    109477       7       34        1.488    1.3694
##    up      68214     68236       4       49        7.481    5.6957
##    up     104260    104274       5       23        4.872    4.5753
##   ...        ...       ...     ...      ...          ...       ...
##    up     108374    108374       6      125        1.506     1.578
##    up      54629     54629       2       14        1.884     1.663
##    up      68262     68262       4       49        7.712     6.407
##    up     108344    108344       6      125        1.536     1.580
##    up     109454    109454       7       34        1.645     1.543
##       meanadult log2FoldChangeadultvsfetal   pvalues significant   qvalues
##       <numeric>                  <numeric> <numeric>    <factor> <numeric>
##    up     1.493                   -0.03085   0.05675       FALSE    0.5639
##    up     1.240                    0.43270   0.06291       FALSE    0.5639
##    up     1.607                    0.23040   0.10867       FALSE    0.5639
##    up     9.267                    0.70220   0.12538       FALSE    0.5639
##    up     5.169                    0.17592   0.16322       FALSE    0.5872
##   ...       ...                        ...       ...         ...       ...
##    up     1.433                    -0.1390    0.7862       FALSE    0.7981
##    up     2.105                     0.3397    0.8055       FALSE    0.7981
##    up     9.017                     0.4930    0.8592       FALSE    0.7981
##    up     1.492                    -0.0830    0.8874       FALSE    0.7981
##    up     1.747                     0.1786    0.9586       FALSE    0.8212
##       significantQval
##              <factor>
##    up           FALSE
##    up           FALSE
##    up           FALSE
##    up           FALSE
##    up           FALSE
##   ...             ...
##    up           FALSE
##    up           FALSE
##    up           FALSE
##    up           FALSE
##    up           FALSE
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
</code></pre>
</div>
<p>The metadata columns are:</p>
<ul>
<li><em>value</em> is the mean F-statistics for the candidate DER.</li>
<li><em>area</em> is the sum of the F-statistics for the candidate DER.</li>
<li><em>indexStart</em> Relates the genomic start coordinate with the filtered genomic index start coordinate.</li>
<li><em>indexEnd</em> Similarly as above but for the end coordinates.</li>
<li><em>cluster</em> The cluster id to which this candidate DER belongs to.</li>
<li><em>clusterL</em> The length of the cluster to which this candidate DER belongs to.</li>
<li><em>meanCoverage</em> The base level mean coverage for the candidate DER.</li>
<li><em>meanfetal</em> In this example, the mean coverage for the fetal samples.</li>
<li><em>meanadult</em> In this example, the mean coverage for the adult samples.</li>
<li><em>log2FoldChangeadultvsfetal</em> In this example, the log2 fold change between adult vs fetal samples.</li>
<li><em>pvalues</em> The p-value for the candidate DER.</li>
<li><em>significant</em> By default, whether the p-value is less than 0.05 or not.</li>
<li><em>qvalues</em> The q-value for the candidate DER calculated with <a href="http://www.bioconductor.org/packages/release/bioc/html/qvalue.html">qvalue</a>.</li>
<li><em>significantQval</em> By default, whether the q-value is less than 0.10 or not.</li>
</ul>
<p>Note that for this type of analysis you might want to try a few coverage cutoffs and/or F-statistic cutoffs. One quick way to evaluate the results is to compare the width of the regions.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Width of potential DERs
summary(width(results$regions$regions))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##    1.00    1.00    3.00    8.48    8.00   39.00
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">sum(width(results$regions$regions) > 50)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] 0
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Width of candidate DERs
sig <- as.logical(results$regions$regions$significant)
summary(width(results$regions$regions[ sig ]))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">sum(width(results$regions$regions[ sig ]) > 50)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] 0
</code></pre>
</div>
<h4>Nearest annotation</h4>
<p><code>analyzeChr()</code> will find the nearest annotation feature using <code>annotateNearest()</code> from <a href="http://www.bioconductor.org/packages/release/bioc/html/bumphunter.html">bumphunter</a>. This information is useful considering that the candidate DERs were identified without relying on annotation. Yet at the end, we are interested to check if they are inside a known exon, upstream a gene, etc.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Nearest annotation
head(results$annotation)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##       name
## 1  C21orf2
## 2    ITSN1
## 3   UBE2G2
## 4     TTC3
## 5 C21orf33
## 6   CRYZL1
##                                                                                                                                                                                                                                                                                                                                                                                        annotation
## 1                                                                                                                                                                                                                 NM_001271440 NM_001271441 NM_001271442 NM_004928 NP_001258369 NP_001258370 NP_001258371 NP_004919 XM_006724051 XM_006724052 XM_006724053 XP_006724114 XP_006724115 XP_006724116
## 2                                                                                                                                                                                                                                                           NM_001001132 NM_003024 NP_001001132 NP_003015 XM_005261026 XM_005261027 XM_005261031 XP_005261083 XP_005261084 XP_005261088 XR_244289
## 3                                                                                                                                                                                                                                                                                                                               NM_001202489 NM_003343 NM_182688 NP_001189418 NP_003334 NP_872630
## 4 NM_001001894 NM_003316 NP_001001894 NP_003307 XM_005261045 XM_005261046 XM_005261047 XM_005261048 XM_005261049 XM_005261050 XM_005261051 XM_005261052 XM_005261053 XM_005261054 XM_005261055 XM_005261056 XM_005261058 XP_005261102 XP_005261103 XP_005261104 XP_005261105 XP_005261106 XP_005261107 XP_005261108 XP_005261109 XP_005261110 XP_005261111 XP_005261112 XP_005261113 XP_005261115
## 5                                                                                                                                                                                                                       NM_004649 NM_198155 NP_004640 NP_937798 XM_005261184 XM_005261185 XM_005261186 XM_006724055 XM_006724056 XP_005261241 XP_005261242 XP_005261243 XP_006724118 XP_006724119
## 6                                                                                                                                                                                                                                                                                                                                                                   NM_145311 NM_145858 NP_665857
##     description region distance     subregion insidedistance exonnumber
## 1   inside exon inside     7485   inside exon              0          5
## 2 inside intron inside     8880 inside intron          -1216          4
## 3   inside exon inside    31648   inside exon              0          4
## 4   inside exon inside    93972   inside exon              0         36
## 5   inside exon inside    11213   inside exon              0          6
## 6          <NA>   <NA>        0          <NA>             NA         NA
##   nexons                         UTR strand geneL codingL
## 1      7 inside transcription region      - 10458    8996
## 2      5 inside transcription region      + 27889   20399
## 3      4                       3'UTR      - 33256    5955
## 4     36                       3'UTR      + 95370   79728
## 5      6 inside transcription region      + 12111   11251
## 6     10                        <NA>      -     0       0
</code></pre>
</div>
<p>For more details on the output please check the <a href="http://www.bioconductor.org/packages/release/bioc/html/bumphunter.html">bumphunter</a> package.</p>
<h4>Time spent</h4>
<p>The final piece is the wallclock time spent during each of the steps in <code>analyzeChr()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Time spent
results$timeinfo</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##                      init                     setup 
## "2014-09-30 14:46:10 EDT" "2014-09-30 14:46:10 EDT" 
##                  prepData                  savePrep 
## "2014-09-30 14:46:15 EDT" "2014-09-30 14:46:15 EDT" 
##            calculateStats                 saveStats 
## "2014-09-30 14:46:16 EDT" "2014-09-30 14:46:16 EDT" 
##             saveStatsOpts          calculatePvalues 
## "2014-09-30 14:46:16 EDT" "2014-09-30 14:46:24 EDT" 
##                  saveRegs                  annotate 
## "2014-09-30 14:46:24 EDT" "2014-09-30 14:46:25 EDT" 
##                  saveAnno 
## "2014-09-30 14:46:25 EDT"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Use this information to make a plot
timed <- diff(results$timeinfo)
timed.df <- data.frame(Seconds = as.numeric(timed), Step = factor(names(timed),
    levels = rev(names(timed))))
library('ggplot2')
ggplot(timed.df, aes(y = Step, x = Seconds)) + geom_point()</code></pre>
<div class="row">
<div class="col-md-offset-3 col-md-6">
<a href="#" class="thumbnail"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfgAAAH4CAYAAACmKP9/AAAEJGlDQ1BJQ0MgUHJvZmlsZQAAOBGFVd9v21QUPolvUqQWPyBYR4eKxa9VU1u5GxqtxgZJk6XtShal6dgqJOQ6N4mpGwfb6baqT3uBNwb8AUDZAw9IPCENBmJ72fbAtElThyqqSUh76MQPISbtBVXhu3ZiJ1PEXPX6yznfOec7517bRD1fabWaGVWIlquunc8klZOnFpSeTYrSs9RLA9Sr6U4tkcvNEi7BFffO6+EdigjL7ZHu/k72I796i9zRiSJPwG4VHX0Z+AxRzNRrtksUvwf7+Gm3BtzzHPDTNgQCqwKXfZwSeNHHJz1OIT8JjtAq6xWtCLwGPLzYZi+3YV8DGMiT4VVuG7oiZpGzrZJhcs/hL49xtzH/Dy6bdfTsXYNY+5yluWO4D4neK/ZUvok/17X0HPBLsF+vuUlhfwX4j/rSfAJ4H1H0qZJ9dN7nR19frRTeBt4Fe9FwpwtN+2p1MXscGLHR9SXrmMgjONd1ZxKzpBeA71b4tNhj6JGoyFNp4GHgwUp9qplfmnFW5oTdy7NamcwCI49kv6fN5IAHgD+0rbyoBc3SOjczohbyS1drbq6pQdqumllRC/0ymTtej8gpbbuVwpQfyw66dqEZyxZKxtHpJn+tZnpnEdrYBbueF9qQn93S7HQGGHnYP7w6L+YGHNtd1FJitqPAR+hERCNOFi1i1alKO6RQnjKUxL1GNjwlMsiEhcPLYTEiT9ISbN15OY/jx4SMshe9LaJRpTvHr3C/ybFYP1PZAfwfYrPsMBtnE6SwN9ib7AhLwTrBDgUKcm06FSrTfSj187xPdVQWOk5Q8vxAfSiIUc7Z7xr6zY/+hpqwSyv0I0/QMTRb7RMgBxNodTfSPqdraz/sDjzKBrv4zu2+a2t0/HHzjd2Lbcc2sG7GtsL42K+xLfxtUgI7YHqKlqHK8HbCCXgjHT1cAdMlDetv4FnQ2lLasaOl6vmB0CMmwT/IPszSueHQqv6i/qluqF+oF9TfO2qEGTumJH0qfSv9KH0nfS/9TIp0Wboi/SRdlb6RLgU5u++9nyXYe69fYRPdil1o1WufNSdTTsp75BfllPy8/LI8G7AUuV8ek6fkvfDsCfbNDP0dvRh0CrNqTbV7LfEEGDQPJQadBtfGVMWEq3QWWdufk6ZSNsjG2PQjp3ZcnOWWing6noonSInvi0/Ex+IzAreevPhe+CawpgP1/pMTMDo64G0sTCXIM+KdOnFWRfQKdJvQzV1+Bt8OokmrdtY2yhVX2a+qrykJfMq4Ml3VR4cVzTQVz+UoNne4vcKLoyS+gyKO6EHe+75Fdt0Mbe5bRIf/wjvrVmhbqBN97RD1vxrahvBOfOYzoosH9bq94uejSOQGkVM6sN/7HelL4t10t9F4gPdVzydEOx83Gv+uNxo7XyL/FtFl8z9ZAHF4bBsrEwAAQABJREFUeAHt3QmYFNW5//F3mGHfV9kXWSVCMIASQSAaQSMBFQTFq+BKlHBNXPKA8eYuMWjUxOfRhxBEuAQXDGCMYhQlgOyKIAKCsoqy77LIvvz9nfuvTk8zA9PT1TVVzfc8z0x313LOqU+d6fecUzXdWae/S0ZCAAEEEEAAgYwSKJZRR8PBIIAAAggggIATIMDTEBBAAAEEEMhAAQJ8Bp5UDgkBBBBAAAECPG0AAQQQQACBDBQgwGfgSeWQEEAAAQQQIMDTBhBAAAEEEMhAAQJ8Bp5UDgkBBBBAAAECPG0AAQQQQACBDBQgwGfgSeWQEEAAAQQQIMDTBhBAAAEEEMhAAQJ8Bp5UDgkBBBBAAAECPG0AAQQQQACBDBQgwGfgSeWQEEAAAQQQyIEgnAJ79+61kydPplS54sWL24kTJ4wvDEyJMamdixUrZtnZ2Xb8+PGk9mPj1ARo66n5FWZv2nph1JLfp1SpUlauXLnkd/xuDwJ8odjSv9PBgwdTDvC1a9e23bt329GjR9NfYZ9LUKM+cuSIz7mmP7syZcpY6dKlTR20qKWcnBzXGUy1Y1kUx01bD149ym1dHcJTp06l/B4blHphAzxT9EGdIcpBAAEEEEAgQAECfIDYFIUAAggggEBQAgT4oKQpBwEEEEAAgQAFCPABYlMUAggggAACQQkQ4IOSphwEEEAAAQQCFCDAB4hNUQgggAACCAQlQIAPSppyEEAAAQQQCFCAAB8gNkUhgAACCCAQlAABPihpykEAAQQQQCBAAQJ8gNgUhQACCCCAQFACBPigpCkHAQQQQACBAAUI8AFiUxQCCCCAAAJBCRDgg5KmHAQQQAABBAIUIMAHiE1RCCCAAAIIBCVAgA9KmnIQQAABBBAIUIAAHyA2RSGAAAIIIBCUAAE+KGnKQQABBBBAIEABAnyA2BSFAAIIIIBAUAIE+KCkKQcBBBBAAIEABQjwAWJTFAIIIIAAAkEJEOCDkqYcBBBAAAEEAhQgwAeITVEIIIAAAggEJUCA90l6+fLlduDAAZ9y8yebb7/91k6fPu1PZuSCAAIIIBApAQJ8AU7X0aNH7dSpU7m21Ov44Dl37lw7duxYrm3yenH8+PG8Fvu6TPX65S9/aVWqVLG2bdva0qVLfc2fzBBAAAEEwi+Q9V0wYIh3lvM0fvx4U4Dft2+f/fjHP7Y2bdrYzJkzbd26dXbixAnr0KGDVapUyUaPHm0NGza0G264wV5//XUbNGiQy3X48OH26KOP2jPPPGOlS5e2YsWKWYUKFax///6WlZXltjl06JDLM74al1xyiZUpUyZ+UYGfT58+3fr06RPb/oorrrC33nor9joKT3JycpxvFOoaX8fixYtbiRIlTLMnUUtqm3o7iOJbQvny5U1/RydPnowau9HWgz9lUWrr+nusXLlyoZByCrXXebKTRunr16+3+++/3wXbrVu3ujeRlStX2uDBg92ofsSIETZkyBCrW7eu3Xjjje7NMX607z0/ePCg9evXz+rVq2cvvviirV692po3b+4kdQLViUhMXgcgcfm5Xie+yel1YfM6V1npWq/6Rq3OnkVU6+7V22uz3vFE5dGrf1Tq69UzyvXWMaj+UUtegI9CW0+lw02AP0vLVCPQiHzChAluRNa9e3cX4Hfv3m2jRo1ye2ZnZ9v+/fvzzcVrQBrZqROgVL9+fdu5c2cswJctW9Z1DuIz2bhxo+3duzd+UYGfa1r+6quvtmnTprnZhV/84heFzqvAhfq8YalSpezIkSM+55r+7DTronZT2HOX/hrmX4JGknozSewg5r9HeNZodkz3wOTVUQ5PLfOuSZTbumZOotjW9X6s9+YotPVy5crl3XAKsJQAfxYkXVPftWuXPfDAA6agrtG6gmXFihXdFLwayHvvvWcK0OrF6s1Rb5KaKlTStL73hqNr79u3b7eaNWvahg0b7MorrzxLyamtUqfjhRdeMD2qjqn0AFOrCXsjgAACCBSVAAH+LPK6lqreqabUDx8+bJ07d3bXz1u0aOECqIJn06ZNXSDV6HzMmDH24IMPumvtI0eOtJIlS7rtVYRGdRMnTnTbVq9e3Zo0aXKWkv1ZdcEFF7iOidfJ8CdXckEAAQQQiIIAAf4cZ0k3q2n0rRG6RudKmqr3pnY0Slbq0aOHW66pH12zV4dA04ZeUoDXtXrNCijwkxBAAAEEEEinAAG+ALoK2onJC+zxy+O3iw/u2sabxie4x4vxHAEEEEAgXQL8H3y6ZBPy1f+lkxBAAAEEEAhKgAAflDTlIIAAAgggEKAAAT5AbIpCAAEEEEAgKAECfFDSlIMAAggggECAAgT4ALEpCgEEEEAAgaAECPBBSVMOAggggAACAQoQ4APEpigEEEAAAQSCEiDAByVNOQgggAACCAQoQIAPEJuiEEAAAQQQCEqAAB+UNOUggAACCCAQoAABPkBsikIAAQQQQCAoAQJ8UNKUgwACCCCAQIACBPgAsSkKAQQQQACBoAQI8EFJUw4CCCCAAAIBChDgA8SmKAQQQAABBIISIMAHJU05CCCAAAIIBChAgA8Qm6IQQAABBBAISoAAH5Q05SCAAAIIIBCgAAE+QGyKQgABBBBAICgBAnxQ0pSDAAIIIIBAgAIE+ACxKQoBBBBAAIGgBAjwQUlTDgIIIIAAAgEKEOADxKYoBBBAAAEEghIgwAclTTkIIIAAAggEKECADxCbohBAAAEEEAhKgAAflDTlIIAAAgggEKAAAT5AbIpCAAEEEEAgKAECfFDSlIMAAggggECAAgT4ALEpCgEEEEAAgaAECPBBSVMOAggggAACAQoQ4APEpigEEEAAAQSCEiDAByVNOQgggAACCAQoQIAPEJuiEEAAAQQQCEqAAB+UNOUggAACCCAQoAABPkBsikIAAQQQQCAogZygCsqEcmbMmGELFiyw7OxsdzhNmjSxHj16WKlSpfI8vMOHD9sXX3xhl1xySZ7rWYgAAggggEC6BM6bEfzx48fPMDxx4kRs2cmTJ2PPvSenTp2y+OUK2F27drWhQ4faww8/bCVLlrRJkyZ5m5u2P3r0aOz19u3b7bPPPou9TlwfW8ETBBBAIA0CW7dutV//+tf20EMP2Zo1a9JQAlmGWSDjR/DPPPOMlS5d2ooVK2YVKlSw/v3724QJE2zv3r125MgRGzJkiP3973+3/fv32+nTp+26665z248aNco9alm7du2sU6dOuc5jTk6O/eQnP3HBXp2AOXPm2LJly9zoXh0H5Tt37lxbu3atW75nz54z1qtOStr/m2++yZW/XqiMVFJWVparT6r5pFKHwu6rWZKo1lvuUa272rvqH7VEW8/7jN17772xQYbej+bPn+8GJnlvndxS/Y1Gua2r7lFs68mcpdQiSDIlFdG2Bw8etH79+lm9evXsxRdftNWrV7uAqmnzyy+/3DX+8uXLW9++fW3nzp02bdo0u/baa10H4JFHHnEN4De/+Y1ddtllZxyBGnjVqlXdtnpjHDx4sOskPPvss7Zr1y7XKdDy1q1b28yZM89YX6NGDZfnvn377Pe//32u/B988EGrXr16rmXJvlAHQvWLYtIfnuyimOReq1atyFXde7OLojtt/czmpoHGihUrYit27Njh3vv8aptegPQrv1hFA3gSpbb+7bffFlok4wN88eLFrW7dug6ofv36LojrRc2aNd0yXSP/6quv7Ouvv3avNdpXatCgQexauxqwRuB5pQMHDpg6CHqDGTlypJUpU8YOHTqUa2pf+51tfZUqVeypp57Klf3GjRtNP6mk2rVr2+7du3NdNkglvyD31X0NmmGJWtL5V3vQ5ZmoJc06KLjHX5aKyjHQ1vM+U1deeaVNnz7drbzooovczFKq7yteSVFu64oLiZdgveMK22O5cuUKXaWMD/C69q43WwX0DRs2mBr8l19+6QKu1HSjnIK6puY1kv70008d5qZNm9ybnd7wtmzZkudI+MMPP3SdB/WUZ82aZY899pjb94knnnD7Kqhrf/XA8lrvNuYXAgggkCaBESNG2OTJk10nv0+fPrH3vTQVR7YhE8j4AK8gO3HiRDca15S3ArruhPdSy5Yt3TT92LFjXSDu1q2bW6Xp9+eee84UvHv27Bm7pjp16lR3vV2jHE1/33rrrVa2bFk3Jfvyyy+70Y9eq7OgywLr1693nQbNAiSu16iDhAACCKRLQDcC6z2KdH4KZH03wozmhc4Cnq/HH3/c3UV67Nixs95covUlSpRwueoGvFdffdV+9rOfuRF4QW+Y0pRy4r/MaRpIxOow5LU+v8PQNFqqU6VMW+anm77lUZ62ZIo+fe3ibDlzOepsOulZF7Up+mrVqhUKIuNH8BpN64YK9WTPlrzgrm20vd6oFZSTSYnBXftqBsFLea331vGIAAIIIICAnwL/ij5+5hqivH75y18mXZtKlSrZHXfckfR+7IAAAggggEBYBDI+wIcFmnoggAACCCAQpAABPkhtykIAAQQQQCAgAQJ8QNAUgwACCCCAQJACBPggtSkLAQQQQACBgAQI8AFBUwwCCCCAAAJBChDgg9SmLAQQQAABBAISIMAHBE0xCCCAAAIIBClAgA9Sm7IQQAABBBAISIAAHxA0xSCAAAIIIBCkAAE+SG3KQgABBBBAICABAnxA0BSDAAIIIIBAkAIE+CC1KQsBBBBAAIGABAjwAUFTDAIIIIAAAkEKEOCD1KYsBBBAAAEEAhIgwAcETTEIIIAAAggEKUCAD1KbshBAAAEEEAhIgAAfEDTFIIAAAgggEKQAAT5IbcpCAAEEEEAgIAECfEDQFIMAAggggECQAgT4ILUpCwEEEEAAgYAECPABQVMMAggggAACQQoQ4IPUpiwEEEAAAQQCEiDABwRNMQgggAACCAQpQIAPUpuyEEAAAQQQCEiAAB8QNMUggAACCCAQpAABPkhtykIAAQQQQCAgAQJ8QNAUgwACCCCAQJACBPggtSkLAQQQQACBgAQI8AFBUwwCCCCAAAJBChDgg9SmLAQQQAABBAISIMAHBE0xCCCAAAIIBClAgA9Sm7IQQAABBBAISIAAHxA0xSCAAAIIIBCkAAE+SG3KQgABBBBAICABAnxA0BSDAAIIIIBAkAI5QRYW9bJmzJhhCxYssOzsbHcoNWrUsJtuusnKly8f9UOL1X/79u22YsUKa9OmjVWpUiW2nCcIIIAAAtESyPgAf+TIEStVqlSus3LixAk7ffq0FS9ePNfyxBenTp2yrKws96N1hw8ftq5du1rHjh3dpu+9956988471q9fP/da+ebk/B9p4r5ug+9+HT9+/JzletsG/bh06VK75ZZb3HFWrFjR3njjDWvUqFHQ1aA8BBBAAAEfBDI2wB86dMheeOEFq1Chgu3Zs8duv/12q169uv3lL39xQXbv3r3WvHlza9u2rU2dOtXuvvtux/nUU0/ZAw88YPPnz7d169aZgnaHDh3ciDbRu2rVqvbNN9/YRx99ZB9++KHt37/f7rnnHvv888/P2PeZZ56x0qVLW7FixVyd+vfvH+s47Nu3z8aNG5cr+759+1q1atVyLUv2hTobGoWrs1GQ9Oabb7rgrm1VJ3Vg/vM//7Mgu/q+jZwKWm/fC08hQ9Vb7hdccEEKuRTNrurMKqnzG7WUbFsP0/FFta1rJlM/UWzrMlc7j0JbP3bsWKGba8YG+C1btljJkiVNgVKBV0lBv0WLFi5gHzhwwJ5++mnr1auXKdhr3Y4dO6xWrVp28uRJW7lypQ0ePNgFmREjRsQC/MKFC+3rr792eW7dutUF9I0bN1rlypVdx0D55LXvwYMH3Ui/Xr169uKLL9rq1atdB0P1UuDv+t3MQHxScFMdU0maoVB9NGtQkKQOUHxSByPVOsTnl8xz1b2g9U4m33Rvqzan81lUbqkcn970lKLYsUq2rafi5Pe+tHW/Rc+dnzomCu5RaOverPC5j+rMLTI2wF944YXWpEkTe/75590bbp8+fdxoVsF58eLFbhTt9d7at29vS5Yssc2bN7vgv2bNGtu9e7eNGjXKiakxeJ2Exo0bu2BfokQJ0whef5wK8F4vNr99tV3dunVdfvXr17edO3fGArzy+v73v5/r7ChPBedUUqVKlUyXKI4ePVqgbAYOHOhmHhYtWmRdunSx66+/PuU6FKjgPDbSZRXVPYpJ5zPVc1cUx603Ev1NqIMbtZRsWw/T8dHWgz8bej9WcI9CWy9XrlyhgTI2wG/bts10E9ywYcPcFPrs2bOtZcuWpukOjcw1Bf3EE084uHbt2tm476bIFVB005xGX7oGPWjQINcINFVdtmxZt62CugJ0YvJGP7pmnde+Go3qBraaNWvahg0b7Morr0zMoshfa+SpSwkkBBBAAIHoC2RsgNd0s64pf/LJJ+4avEajCvjTp0+3V155xfXcypQp40a36iGpF63Rua5D6rq9pvJ1DV+9vKZNm8bunD/XKc9vX3UAJk6c6PJR3TS7QEIAAQQQQCBdAlnfTclF746aJDQ0Pa0pU+8GIu2qu+E1Wj1X8qZvNEWfbErc9/HHH7df//rXbgZB12nPlTRF7+Vxrm3zW1+7dm13qaGgU/T55VMUy6M6balOo/5tUrM1UUtRnqKnrQff2qLc1qM2RV/YG64zdgTvNfe8gmlBgrv2L0xg98pN3FdT/Opk5FUfbx8eEUAAAQQQ8Evg/26b9Ss38slX4Je//GW+61iBAAIIIICA3wIEeL9FyQ8BBBBAAIEQCBDgQ3ASqAICCCCAAAJ+CxDg/RYlPwQQQAABBEIgQIAPwUmgCggggAACCPgtQID3W5T8EEAAAQQQCIEAAT4EJ4EqIIAAAggg4LcAAd5vUfJDAAEEEEAgBAIE+BCcBKqAAAIIIICA3wIEeL9FyQ8BBBBAAIEQCBDgQ3ASqAICCCCAAAJ+CxDg/RYlPwQQQAABBEIgQIAPwUmgCggggAACCPgtQID3W5T8EEAAAQQQCIEAAT4EJ4EqIIAAAggg4LcAAd5vUfJDAAEEEEAgBAIE+BCcBKqAAAIIIICA3wIEeL9FyQ8BBBBAAIEQCBDgQ3ASqAICCCCAAAJ+CxDg/RYlPwQQQAABBEIgQIAPwUmgCggggAACCPgtQID3W5T8EEAAAQQQCIEAAT4EJ4EqIIAAAggg4LcAAd5vUfJDAAEEEEAgBAIE+BCcBKqAAAIIIICA3wIEeL9FyQ8BBBBAAIEQCBDgQ3ASqAICCCCAAAJ+CxDg/RYlPwQQQAABBEIgQIAPwUmgCggggAACCPgtQID3W5T8EEAAAQQQCIEAAT4EJ4EqIIAAAggg4LcAAd5vUfJDAAEEEEAgBAIE+BCcBKqAAAIIIICA3wIEeL9FyQ8BBBBAAIEQCBDgQ3ASqAICCCCAAAJ+C0QywO/Zs8deeeWVpCyef/75fLc/fPiwLVmyJN/1WjFt2jT73e9+Z8OHD7dx48aZ9lFavny5HThwwD3P69e51ue1D8sQQAABBBBIVSCUAf7EiRN2+vTpXMd29OhRO3XqlFumx/3798fWa1vtc7a0b9++XKuPHDkSe719+3b77LPPYq/15OTJk7HXmzdvduuHDh1qw4YNs+rVq9s///lPt37u3Ll27Nix2Lbx9dTC+PWqt9ZnWlJnZ8iQIda2bVv36HV+Mu04OR4EEEAgSgI5YavslClTbN26dVa8eHHr2rWr1a1b18aPH2+lS5e2rVu3Ws+ePa1OnTqxai9dutSmTp1qVapUsXr16lm7du3s9ddft0GDBrltNOJ+9NFHY9uvWbPG3n77bStXrpzt2LHD7r33XheE165da8uWLbNatWrZO++8Y8ePH7dSpUpZnz59rGTJkm6Urm0aN25s3bt3d+s3bNhgmzZtsr/97W/Wt2/fM+pZsWLF2Pqrr77aJk+ebJUqVTJ1Lu677z7Lzs529VLQ//TTT2N11BMdt8pPJRUrVsy55eSk9zS/+OKLzlR1le0ll1ziAn0qddf593xSySfofUuUKOHqXbZs2aCLTrk8eauz7HWkU84wwAzU1vX3ku62no5Dimpb1/ui2kwU27raidp5FNq62nZhU3rf+ZOs1ZYtW0wB+MEHH7RDhw7ZokWLrHz58i6gNmvWzBTMNZUeH+AnTpxov/nNb1wQ1jS6Rt7xJy3+uapz8OBBGzBggFWuXNneeOMNW716tXXq1Mm9sbVu3dot69ixozVp0sTmz5/vpuDbt29v/fv3t1mzZrkgriDfu3dva9iwoQvEN954o2mGQIE/vp4DBw6MrV+4cKG1bNnSFOjVgdGoX50WJQV4HVt8ql+/fsp/ON6bnoJOOlPi7Ihep/pHr7qnu97pMNEbXlTf9LKyshxJ4uxZOpz8zlPtRX9PiX/vfpeTjvxo6+lQPXueUWrr8TPEZz+qM9eGKsDv3LnTGjRo4GpZpkwZ69y5s+l6+/vvv2/Tp093o3r9MXhJnYAKFSq44K5lCp67du3yVrvHxD945TthwgTX09dUsqbb49OqVatMU/YqT290CvS7d++2GjVq2F133eU6EB999JEL9JqW9pJmBPKrp7bp0qWLG91qRkHBWz9e0jFoJiE+bdy40c0wxC9L9nnt2rVt7969ab8scP3119urr77qzpVmUvRasyOpJI3G4i+jpJJXkPuqfalTmurxB1lnryyNatTm4y9PeevC/hhUW0+HA209Hapnz1OzJooNUWjrii2FTaEK8Aqms2fPdsei4P3WW2+50VCrVq3ssssus8WLF+e6GU5vpnpTUg9Hoz0FmR//+Mdu9K9MNJJMvOat6fuf/exnbkr/tddecydZnQZv1KLRuUbyzZs3N03Ja6r+yy+/tJUrV9ptt93m6tO0aVNTkFdST1D7zpgxw/Kqp7deMxOaKdCUv8pdsWKFaWYgE5LMZs6caeocyU0dFhICCCCAQNEKhCrAa1pXU9yjR492N811/e4avAKkrrErQCqYx99cJ7pu3brZ2LFj3TqNijXS1lTdyJEj3cg+MdjoGr2m9dU5UPDXo0ad69evt3nz5tkVV1zhRtp6rk7G7bff7qabtV6jb/WmtN/NN9/szpyulY8ZM8Z69Ojh7rRPrKe3XttPmjTJBT8dg7bPpCTnTOmwZNJ54VgQQOD8Fcj6bvSZ+3b1EFjojniNqr3peE2laNnZrslqpK1pFy9p+t27xu0t8x414veulXrLVIYotFzJmxXw1utR22jaWJ2C+OSVnV89vfXa52z1is9TU/SpTh9p2lKXFxJnMeLLCevzqE9b6jJP1FLUp+hp68G2OL0P6nJUFNt61Kboq1WrVqiTG6oRvHcEeqOJTwr0Zwvu2jY+uOt1fsFd6/LKy+tMaL1SftskBndt65WdXz299dr2bPXSehICCCCAAAJ+CPzrjjU/ciMPBBBAAAEEEAiFAAE+FKeBSiCAAAIIIOCvAAHeX09yQwABBBBAIBQCBPhQnAYqgQACCCCAgL8CBHh/PckNAQQQQACBUAgQ4ENxGqgEAggggAAC/goQ4P31JDcEEEAAAQRCIUCAD8VpoBIIIIAAAgj4K0CA99eT3BBAAAEEEAiFAAE+FKeBSiCAAAIIIOCvAAHeX09yQwABBBBAIBQCBPhQnAYqgQACCCCAgL8CBHh/PckNAQQQQACBUAgQ4ENxGqgEAggggAAC/goQ4P31JDcEEEAAAQRCIUCAD8VpoBIIIIAAAgj4K0CA99eT3BBAAAEEEAiFAAE+FKeBSiCAAAIIIOCvAAHeX09yQwABBBBAIBQCBPhQnAYqgQACCCCAgL8CBHh/PckNAQQQQACBUAgQ4ENxGqgEAggggAAC/goQ4P31JDcEEEAAAQRCIUCAD8VpoBIIIIAAAgj4K0CA99eT3BBAAAEEEAiFAAE+FKeBSiCAAAIIIOCvAAHeX09yQwABBBBAIBQCBPhQnAYqgQACCCCAgL8CBHh/PckNAQQQQACBUAgQ4ENxGqgEAggggAAC/goQ4P31JDcEEEAAAQRCIUCAD8VpoBIIIIAAAgj4K0CA99eT3BBAAAEEEAiFAAE+FKeBSiCAAAIIIOCvAAHeX09yQwABBBBAIBQCOaGoRQQqMW3aNFu4cKFlZWVZ7dq1rV+/fla6dGlbvny5NWzY0MqXL5/nUZxrfZ47sRABBBBAAIEUBTJ2BH/8+HE7ceJELp5Tp07Z0aNHcy3L78XJkydjqzZv3myfffaZDR061IYNG2bVq1e3f/7zn2793Llz7dixY7Ftlb/K8VL8+mTK9/b341F1Gjt2rP3xj380HQsJAQQQQCDzBTJyBD979mxbsmSJ5eTkWN26da1Xr172wQcf2LJlyyw7O9sF/iFDhrigd80117htFi1aZDt37rR27drZO++8Y+oglCpVyvr06WMlS5a0AwcO2Nq1a61x48bWvXt3t37Dhg22adMm+9vf/mZ9+/a18ePHu1H91q1brWfPnlaxYsXY+quvvtomT55slSpVsiNHjth9993n6hJEE3vooYfsH//4hytq4sSJNmPGDCtTpkwQRVMGAggggEARCWRkgF+6dKl169bNmjdv7oK6bE+fPm2DBw92j88++6zt2rXL2rdvbwrs6gR8/PHHdtNNN9mcOXOsY8eO1qRJE5s/f76bgtd2/fv3t1mzZrkgriDfu3dvNzWvfW+88Ubbt2+fC/zNmjUzla8OxsCBA13eWq/p/ZYtW5oC/bp169yoX1P8SqrLf//3f7vn3q/HHnvMatSo4b0s9GPNmjXdMXkZbN++3XVWLrroIm8Rj2kQaNCgQRpyJcuzCaitk4IXoK2n1/zgwYOFLiAjA7yC77vvvmsarf7gBz+wNm3aWLFixWzkyJFu5Hro0CHTFPzFF19sU6dOdcFZ0/nVqlWzVatWmYLg9OnTXWdAgX737t0u2N51111uv48++sgFes0CeKlcuXL2/vvvu/2KFy/uyvPW6bFLly729ttv2/Dhw61+/frux1tfpUoV+/3vf++9dI8qc+PGjbmWJftCb3h79+51HRkdj1LlypVNdU0172Trkuz2mjUp6OWUZPNO5/bqtOl+jB07dqSzmLTkrdktdYTjLzGlpaA0ZKq2vmfPnlyXy9JQTFqyjGpb1yxg2bJl3cxnWmDSmKlmd9XOo9DW5ZzfPV7nIsrIAP/555/bbbfd5o796aeftrZt27rRt0bFSk888YR7I9MbWtOmTd3U+aWXXurWaXTeunVrN/rXlLym6r/88ktbuXKly9PbR0FeSTfd6U1R096tWrWyyy67zBYvXuxG8PHr16xZY506dXJT/q+99pqtWLHCBV5to85H4pS5ArwfjU956Nr76NGj7ZtvvrHbb7/d/VH6kbfqnq4U1UCjeke17mqHUa272mFU6x7VenvvId5jut4L0pGvzKPqnoxHRgZ49XYU0NQzrlOnjrvrvVatWvbyyy+7Ebh6nZpS193wHTp0cAHw3/7t35zbFVdc4Uba8+bNM430vYC4fv16N/rW6Fcjy5tvvtltryn6MWPGWI8ePUx32iuQq3e4f//+XOu1/aRJk6xChQpunbYPKqlMXYcnIYAAAgicPwJZ3/ViTmfi4eqwdHe7gryXdHObbpwrSNK+JUqUyLWpeqrKI3G0rVG+puW1XlP9ift565XZ4cOH3Y14uTLO44Wm0OPv5M9jk3MuUgdGMwFRnOrWeZJ11JI3nabLPFFL6pjq7ybVdlcUx01bD149ym3de7+OQlvXoFKXjwuTMvbf5DR1Hh/chVPQ4K5tE4O0luU1la7laixKWp/Xft56bePdWKfnJAQQQAABBNIlkLEBPl1g5IsAAggggEAUBAjwUThL1BEBBBBAAIEkBQjwSYKxOQIIIIAAAlEQIMBH4SxRRwQQQAABBJIUIMAnCcbmCCCAAAIIREEgpQCvT0n76quvonCc1BEBBBBAAIHzSqBQAV6f7KZPbNP/5unz1fX417/+9byC42ARQAABBBAIs0ChAvwdd9zhvmBFn7etjz99/fXX3Sel6eNcSQgggAACCCBQ9AKFCvAawf/qV7+yqlWrug950RepDBgwwPQ1rSQEEEAAAQQQKHqBQgV4fTvb2LFj3cey6hD0sZz6DvXLL7+86I+IGiCAAAIIIICAFSrA63Pa7777bjeC11eu6otc9DWrffv2tRYtWtg999wDLQIIIIAAAggUoUChvk3uySeftP/5n/9xI3h9Gb0+DF9fVOGlihUrek95RAABBBBAAIEiEPhXVE6icAXwxx9/3N566y0bMmSIm6LX16YOHTo0iVzYFAEEEEAAAQTSJVCoKfo777zTLrzwQjeKV8V0w92rr75qq1evTlc9yRcBBBBAAAEEkhBIOsDr+6JXrFhhjzzySOzrV+vXr2/9+vWzGTNmJFE0myKAAAIIIIBAugSSDvD6nnVdc1++fHmsTqdOnbK3337b3WwXW8gTBBBAAAEEECgygUJdg3/qqaesc+fO1qhRI/d/8KNGjXKfaNejR48iOxAKRgABBBBAAIF/CRQqwPfu3dv073H/+Mc/7Pjx4+5T7Zo2bfqvXHmGAAIIIIAAAkUqUKgAv3v3bmvevLn78Wqvj63Nzs52/xvvLeMRAQQQQAABBIpGIKlr8EeOHDH9dO3a1T16rw8dOmT//u//blOmTCmao6BUBBBAAAEEEMglkFSAv+6666x06dL22WefuUc9149uups3bx4fVZuLlhcIIIAAAggUnUBSAf6f//ynu+Z+yy23uEddf9fPiRMnbOPGjdasWbOiOxJKRgABBBBAAIGYQFIBXv8ip+vs48ePdx9Nq/+JnzNnjn366aexDHmCAAIIIIAAAkUvkFSAnzRpktWuXduWLFli+t/3nj17uq+JveGGG+zee+8t+qOhBggggAACCCDgBAp8F72+ElbfIDdmzBhr27at/e1vf7Nly5bZF1984Ubz+re5jz/+2Nq3bw8tAggggAACCBSxQIFH8IsXLzZ9D3yfPn2sWLFi9u6779pNN91k5cuXdzfaXXXVVbZo0aIiPhyKRwABBBBAAAEJFDjAHz582N0t77FNnz7dunfv7r20b775xipUqBB7zRMEEEAAAQQQKDqBAgd4TcvPnz/f/Yvc66+/7gK6Ru1KX331lekO+0suuaTojoSSEUAAAQQQQCAmUOBr8A0bNrT/+q//sg4dOridx40bZyVKlLDXXnvN7r//frvvvvvc59HHcuYJAggggAACCBSZQIEDvGr4wAMP2KBBg9w1eAV3pe9973s2a9Ysa9WqlXvNLwQQQAABBBAoeoGkAryqW6pUqVy1JrDn4uAFAggggAACoRAo8DX4UNSWSiCAAAIIIIBAgQQI8AViYiMEEEAAAQSiJUCAj9b5orYIIIAAAggUSIAAXyAmNkIAAQQQQCBaAgT4aJ0vaosAAggggECBBJK+i75AuRbxRkePHrWJEyfa119/bfoGPH0AzzXXXOOe68N6Lr/88nxreK71K1eutLffftt92Y4+pvfHP/6xNW/ePN/8li9fbvoMAW0bZNI3/a1Zs8Z9+1/ifz4EWQ/KQgABBBAoGoHQjOAVlPUNdfFJ3zOv75svSDp58mRsswULFljFihXt17/+tf3qV7+ytWvX2qpVq9z6GTNmxLbTkyNHjuR6Hb/e+657bwN95/1bb71lt9xyiw0dOtRuvfVW06f6abmXEus7d+5cO3bsmFut49NxpjvJ4s477zR9AZA6N9OmTUt3keSPAAIIIBAygVCM4PX98gp8+/btcyPi73//+/aXv/zFBfe9e/e6EbI+Knfq1KnuG+1k+NRTT7kP3lm/fr3NmzfPdQ7q1atn1157rZUpU8ZWrFhhmzZtsrp169o999zjPpznww8/dGW8//771qhRIzcSL1eunO3YscN93e26deti6zXq1dfi5uTkuDx69eplc+bMcZ+/r3KUKlWq5Oqrb9Hbtm2bq4c+AOjgwYPWr18/0yhaddA371199dU2efJkt486Ffrkv+zsbJePtn/zzTfdc+/Xj370I7et9zqZx5kzZ9oHH3zgdlFZzz33nN18883JZFHk28qmbNmyRV6PZCug9lK8eHGrWrVqsrsW+fb6Eim1Wf1ELam9qFMf39GPyjHQ1oM/U1Fq66m06SIP8BrVKkjr424VmLdu3WqHDh2yFi1auI/FPXDggD399NOmAKtgr3UKyLVq1bKSJUu60ak+XU/PX3rpJbfNpZde6gK+Rte7d+92o9gePXq4/PSZ+d26dXPBe8CAAVa5cmV74403bPXq1daxY0f3mfpa//zzz7vtNP2ur8VVUl6Jb9w1atSwhQsXuk5A6dKl3Sf9bdmyxV0i+MUvfuGW33jjjW6bli1bukCvjoRG9dpeSX/gOp7ElDgbkLg+v9eJU/JyLWxe+ZWR7uUKMprBiWJSkI+at5yj9KaX2C689hLFNuPVPfGYwv5alz+j2tb1nqvYE4XOrJwLm4o8wOtN5YYbbrAJEybYt99+60bIGl3r+rm+olbfUOedBH3XvEbVmzdvdsFawV7P9bn4SjphGklrRNyuXTu3jUawCuD69jtdh/eSgp7KVAPVN+VVr17dW+Uee/fu7b4SV9fyf/CDH7ivytU26ozUr18/tq2C9QUXXOBeX3jhhe5RwVqdkfjUpUsXN2MwfPhwt398Hgr0V155Zfzmbtp///79uZYV9EWzZs3cjIRmQVSXxx57zAqbV0HL9Hs7dVISL5/4XUY68lO7UpuKmrcsVG/9raUyYkiHaUHy1Eyc3j+CuARWkPoks02U27oCZRTbumbZFC+i0NbVtgubivwavEayu3btin3O/T/+8Q/3jXVaPnjwYOvZs2dsNKSgrQCv4N+kSRM34lfQ1Uhco3h9X71e67q3RtVK+uNp0KBBLA+vN6TRva6la7+aNWu6k63tvfWff/653XbbbTZs2DA3gtfX4ap85atLCUpatmjRIleuXnvX4tXpqFatmha5/PSmqRveOnXqZP/xH//hZht0CSGdSfVW/XTTIB8nnE5p8kYAAQTCKVDkI3hds9Zo98UXX3Qj6c6dO1vjxo3diPuVV15xPSyNitQzV09GAVvrvUCs76TXCF75qDepb7u77rrrTNf1dW1ePTX9qBOgpOu6+gY8BWuNzr289Ri/XqPx0aNHu2Bcp04ddz1c19w10h4xYoSbXtdMga75N23a1AV+Tc3/6U9/Ml1WUOdASfcAjBkzxl0DnzRpkpuRUI9XlwzSnTQ7QkIAAQQQOD8Fsr4bXYbijhpds1TQ1jShlzR17l2n9pbl96gRv/cNd942ylPTMLo+H5+0XEFf+6hToJ/45K0XjbZJ3F/bavpYnQ0vaWSvEbOm4lUPrwOi9V5+el7QY9JsQKrTR7Vr13b3DTBtKflgkjqK+pfI7du3B1Ogj6VEeYqetu5jQyhgVlFu61GbovdmhAt4amKb/SuaxhYVzROBJ6aCBnftlxjctSyvPOOX57VP/HoF6byCu7aJD+7ePsovr+3j65HMMSlfEgIIIIAAAoURCE2AL0zlw7SP/t+chAACCCCAQFgEuEgbljNBPRBAAAEEEPBRgADvIyZZIYAAAgggEBYBAnxYzgT1QAABBBBAwEcBAryPmGSFAAIIIIBAWAQI8GE5E9QDAQQQQAABHwUI8D5ikhUCCCCAAAJhESDAh+VMUA8EEEAAAQR8FCDA+4hJVggggAACCIRFgAAfljNBPRBAAAEEEPBRgADvIyZZIYAAAgggEBYBAnxYzgT1QAABBBBAwEcBAryPmGSFAAIIIIBAWAQI8GE5E9QDAQQQQAABHwUI8D5ikhUCCCCAAAJhESDAh+VMUA8EEEAAAQR8FCDA+4hJVggggAACCIRFgAAfljNBPRBAAAEEEPBRgADvIyZZIYAAAgggEBYBAnxYzgT1QAABBBBAwEcBAryPmGSFAAIIIIBAWAQI8GE5E9QDAQQQQAABHwUI8D5ikhUCCCCAAAJhESDAh+VMUA8EEEAAAQR8FCDA+4hJVggggAACCIRFgAAfljNBPRBAAAEEEPBRgADvIyZZIYAAAgggEBYBAnxYzgT1QAABBBBAwEcBAryPmGSFAAIIIIBAWAQI8GE5E9QDAQQQQAABHwUI8D5ikhUCCCCAAAJhESDAh+VMUA8EEEAAAQR8FCDA+4hJVggggAACCIRFgAAfljNBPRBAAAEEEPBRgADvIyZZIYAAAgggEBaBnDBU5JNPPrHDhw9bx44dC1SdPXv22Lvvvmu33nprnttv27bNDh06ZBdeeGGe62fMmGELFiyw7Oxst75GjRp20003Wfny5fPcPnHhucpP3J7XCCCAAAIIBC0QyAj++PHjZxzXkSNHYsuOHj3qAry34MSJE3b69Gnv5RmPp06dsv3798eW67Xy8NLy5cttx44d3kv3ePLkydhrdSa6du1qQ4cOdT916tSxd955J7Y+ryfKX+UoJZaf1/ZaFl+mXifWU8tICCCAAAKZJfDpp5/abbfdZgMGDDDFo6JKaR3B79y50/73f//XKleu7EbLAwcOtHXr1tnbb79t5cqVc0H43nvvzXXsU6ZMcdsUL17cBeEDBw640fiVV15pmzZtMo2+f/KTn8T2+eCDD2zZsmUuf3UM7rnnHlu8eLHl5OS4Efzu3btt3rx5LrjWq1fPrr322ti+3pOqVavaN998Y8orKyvLunTpYnv37rVXXnnFnaTx48db6dKlbevWrdazZ09Th8BLw4cPt0cffdS9fPHFF9165aEOgzo2pUqVsj59+phmFSZPnmyVKlUydW7uu+++2AyCtvvqq6+8LN2j9itZsmSuZcm+UD3kGMWkep+tkxfWY1K9ixUrlvK5K4rj04yWzL2ObFHUobBl0tYLK1f4/aLc1hUf1M7T0dYVh+6++25T7FFatWqVaZZabbQwqbD7qay0Bvg333zTBbwWLVrYxx9/bJraPnjwoOvVKOi/8cYbtnr1aveGqMps2bLF1qxZYw8++KAL6osWLXIByjsJeb35aNngwYPdG9Ozzz7r8m/btq2bbtfU+2uvvWaDBg1yb7gvvfSSC9wqa+HChfb111+7mQAFbnUMKlSoYKNHj3YBXmW3a9fO9u3bZ927d7dmzZrZ0qVLbcmSJbkCvFc35emN2OfOnesuNzRp0sTmz5/venC7du2yli1b2tVXX+06MMeOHXOdBu0nk4kTJ+ppLN15551WpUqV2OvCPNEbdsWKFSMZKNWooxjgVW+5q9MYtRRVcznT1oNvberI6oe2ntte7/VecNea7du3W4kSJVx8yb1lwV5pxrmwKa0BXgfaoEEDV7f27du7Rx34hAkT3AhbFa9evXoswGvE721fpkwZ69y5s7tW7h1cXm/4amAjR440ba/r7l6Q1T56vXnzZhs3bpzLQsFYI2mlxo0bW5s2bRy8Gqg30tV1eE3vK5gPGTLEvv32W3v//fdt+vTpbhuVF5/i6+Q9V49NJ1X7aJkCvWYFNHOhEX/9+vXdj5ePOjveLIC3bOPGja7D470uzGPt2rVdQ4u/fFGYfIpiH81gxF/GKYo6FKZMtUO1IZ3/qCWNatRe4/+GonIMtPXgz1SU27re7xUP0tXWr7rqKvf+r7OiWWMN4vRTmKTZ7sKmtAb4pk2bmgKVRr+aWtdNb6+//rr97Gc/c6NTja7jR8AKhLNnz3bHouD81ltvuUCs50qJ19UVfGfNmmWPPfaYW//EE0+4NygFYb1RqQGqA6HrIAoYurFOr9evX+96nQq0ialDhw6mywQ1a9Z0o37VoVWrVnbZZZe5qX+N4OOTytFoXG+O6tAoqfPQunVra968ua1du9ZN1WtmolOnTm66Xse9YsUK8zo98fnxHAEEEEAg2gKjRo1yAV6zYrq8XFQprQFed8UrWGrKWlNoXb+7sU3T3pqOVvDVyNILwgIoW7as6wxomlzXMbS9rndPmzYtNs0RD6Xta9WqZS+//LLriem1ptTVm580aZJbp+l1jeA1RaI6KICfLWkaXTMMmiJXUnCfOnWqu3SgIB5/c5/WK/A/99xzbtSmToTSFVdc4Ubruvavzsntt9/uem+qky4DKI8ePXq4bfmFAAIIIJBZAoo13bp1K/KDyvpuBJr/7eo+VU83kXlT4MpSI14B6CevpOCuUbg3Ha4qqjPgBdDEfTSVm7hOeSh/9aCUVKaCfGGSZhmUX377q27eDSfx+edVpi5L6Ia9cyXNfKQ6fcS05bmU/V8f5WlLpuj9bw8FyVHvXVyOKoiUf9uke4rev5qauyG9WrVqhcoyrSN4r0bxwV3L8guU3vZ6o4lPCtKJATx+fV7rEvM4V5nx+SU+V0fjbPvnd7d7XvsUJLgnls9rBBBAAAEEkhUoluwObI8AAggggAAC4RcgwIf/HFFDBBBAAAEEkhYgwCdNxg4IIIAAAgiEX4AAH/5zRA0RQAABBBBIWoAAnzQZOyCAAAIIIBB+AQJ8+M8RNUQAAQQQQCBpAQJ80mTsgAACCCCAQPgFCPDhP0fUEAEEEEAAgaQFCPBJk7EDAggggAAC4RcgwIf/HFFDBBBAAAEEkhYgwCdNxg4IIIAAAgiEX4AAH/5zRA0RQAABBBBIWoAAnzQZOyCAAAIIIBB+AQJ8+M8RNUQAAQQQQCBpAQJ80mTsgAACCCCAQPgFCPDhP0fUEAEEEEAAgaQFCPBJk7EDAggggAAC4RcgwIf/HFFDBBBAAAEEkhYgwCdNxg4IIIAAAgiEX4AAH/5zRA0RQAABBBBIWoAAnzQZOyCAAAIIIBB+AQJ8+M8RNUQAAQQQQCBpAQJ80mTsgAACCCCAQPgFCPDhP0fUEAEEEEAAgaQFCPBJk7EDAggggAAC4RcgwIf/HFFDBBBAAAEEkhYgwCdNxg4IIIAAAgiEX4AAH/5zRA0RQAABBBBIWoAAnzQZOyCAAAIIIBB+AQJ8+M8RNUQAAQQQQCBpAQJ80mTsgAACCCCAQPgFCPDhP0fUEAEEEEAAgaQFCPBJk7EDAggggAAC4RcgwIf/HFFDBBBAAAEEkhYgwCdNxg4IIIAAAgiEX4AAH/5zRA0RQAABBBBIWiAn6T3Owx3GjRtnu3fvtuzsbDt58qRVqFDBbrrpJqtUqVJoNfbt22fvvfeelS9f3i6//PLQ1pOKIYAAAgikRyBjA/yRI0esVKlSudSOHj1qxYsXt2LFzj5xcerUKcvKynI/ymD//v126623Ws2aNU3rpk+fbu+88471798/lr8CvzoA8SlxmfY9fvy4lSxZMn4z358fOnTIfvrTn9rGjRtd3j//+c/toYce8r0cMkQAAQQQCK9AxgV4BbcXXnjBjbL37Nljt99+uwuo48ePt9KlS9vWrVutZ8+etnfvXhfAu3Tp4p6/8sorpkA4c+ZMW7dunZ04ccI6dOhgbdq0yXX21DlQPl7n4fPPP7d58+a5wF+vXj279tprbdOmTfbyyy9bjRo13Ii/adOm1rBhQ5s8ebIb9avzcd9998U6BKrLc889l6ucQYMGuQ5FroUFfKEOiBfctcvbb79tf/zjHwu4dzg2Uwfr9OnT4ahMErVQvdVG6tSpk8Re4dhUdVeKontOTo5Vr149knWnrQff/qPU1g8fPlxooIwL8Fu2bHEBvW/fvm7kLRlNV3fv3t2aNWtmS5cutSVLlljv3r1t9OjRpgC/aNEia9eunalzsHLlShs8eLAL2CNGjIgF+LFjx7o3bmG3atXKunXr5tCnTZtmCsYalb/00kuus/DGG2+4Eb8CvgK9Ru7qCLRs2dKuvvpq14E4duyY6ygoE02j33HHHS4/75c6GDt37vReJvVYpUoVN1Oh2QKlJk2aFDqvpAr2ceMSJUqYjKKW1PErW7asu6QTtbprBkrBXe01aknBXX/nUWwzUW7rZcqUMQ2kopbUIVQ7j0JbT2XGN+MC/IUXXugC2vPPP+8CaJ8+faxcuXL2/vvvu6l1b4peQVU/O3bscEF/yJAh9sUXX7g35lGjRrn2qjc8Tc8r3XnnnW5E/uabb7rRva7Dq0OwefNm0zV6JTWWbdu2uQZfu3Ztt6xBgwZuWl4dCY2khw8fbvXr13c/boPvfqmxaVl80gi8sG9WmjkYOXKkTZgwwapWrWoPP/xwofOKr1OQzzUKLuzxB1nPxLK8N46o1l0BXpeWopZUb3Voo+ge5bYu9yiaex3ZKLR1dQALmzIuwCvAKsANGzbMPvroI5s9e7YbzWrUfdlll9nixYvdCF5gmoKfMmWKmwpXL6lRo0ZWsWJFNyJXsNZNahqNeUl/iL169XLT3d6oXyOHAQMGuCn7BQsWuGnCxo0b25dffuk6GmvXrjUF+TVr1linTp1MHY7XXnvNVqxYYe3bt/ey9v3xqquusttuu811WHTvAQkBBBBA4PwSyLgAr4CrUfYnn3ziRtLXX3+969lPnTrVBVmNsLxRuabMNcrV6FxJo/IWLVq4a/gK8Lp2nnjjnIK8bq7705/+ZBdddJGb+tcIXr0sbatOgy4HvPvuu/bBBx+4slWm6jVp0iRXhsrv0aPH+dXSOFoEEEAAgUAFsr6bqojenUwFINKoVUHXu5lCAVvXtQsy3eFN2yQG97MVq2kqL29d41fnQJcGFNQvvvhi1xnQ/rqGr5v0zpU0Re/V41zb5rdelwn0731RHMHrWrZuRoxa0jVJXfrZvn171KruLhXp7SDVdlcUB05bD149ym1dl2oVE6LQ1hVHqlWrVqgTnHEjeE8h8cYEjby9AOxtk99jMoHdyyM+b73B6+Y6vVnqf+V1c5+XChLcvW15RAABBBBAoLACGRvgCwvix366a10/JAQQQAABBIpK4Oyf+FJUtaJcBBBAAAEEEEhJgACfEh87I4AAAgggEE4BAnw4zwu1QgABBBBAICUBAnxKfOyMAAIIIIBAOAUI8OE8L9QKAQQQQACBlAQI8CnxsTMCCCCAAALhFCDAh/O8UCsEEEAAAQRSEiDAp8THzggggAACCIRTgAAfzvNCrRBAAAEEEEhJgACfEh87I4AAAgggEE4BAnw4zwu1QgABBBBAICUBAnxKfOyMAAIIIIBAOAUI8OE8L9QKAQQQQACBlAQI8CnxsTMCCCCAAALhFCDAh/O8UCsEEEAAAQRSEiDAp8THzggggAACCIRTgAAfzvNCrRBAAAEEEEhJgACfEh87I4AAAgggEE4BAnw4zwu1QgABBBBAICUBAnxKfOyMAAIIIIBAOAUI8OE8L9QKAQQQQACBlAQI8CnxsTMCCCCAAALhFCDAh/O8UCsEEEAAAQRSEiDAp8THzggggAACCIRTgAAfzvNCrRBAAAEEEEhJgACfEh87I4AAAgggEE4BAnw4zwu1QgABBBBAICUBAnxKfOyMAAIIIIBAOAUI8OE8L9QKAQQQQACBlAQI8CnxsTMCCCCAAALhFCDAh/O8UCsEEEAAAQRSEiDAp8THzggggAACCIRTgAAfzvNCrRBAAAEEEEhJgACfEh87I4AAAgggEE4BAnwK5+Xw4cO2ZMmSfHM41/p8d2RF2gSOHTuWtrzJGAEEEAiTwHkV4E+dOmVHjx49w//kyZOmn8SUVzCI32779u322WefxXZLzD9xvTaM3z+2I08CEfjtb39rLVu2tEsvvdQWLlwYSJkUggACCBSVQE5RFRx0uR988IEtW7bMsrOz7cSJEzZkyBBbvHixLVq0yBSY9+7dazfccIPVrl3bxowZY6VKlbIjR47YRRddZNddd50tX77cZs6caSVKlLA6derYT3/6U5s7d66tXbvW5btnz54z8o9fX7x4cZs3b54rq169enbttdfGCA4dOmRz5syJvdaTiy++2MqVK5drWbIvdKzKQ8cStZSTk2MlS5b0rdqaaRk7dqzLb+fOnfa73/3O1Cb8TjrPqnvFihX9zjrt+RUrVsxOnz7tftJemMTDA6kAABLGSURBVM8FqK2XLVuWtu6z69myi3JbV3vR+77aeyan8ybA60QOHjzYndBnn33Wdu3aFTu5Wq5ArYDcq1cv08j7ySefdOddoz4F+OnTp7v91ahHjx7ttunUqZPLo3Xr1i74J+Yfv/65556zQYMGuaD10ksvuQ5F5cqVXRmq2759+3K1s6ysLNcZybWwEC/0pq3GHLXkd73VqYtPmp1Jh4vq7de5i69vEM+jHODlk47zGZR7FOse9bauv9MoBHh1RAqbzpsAr8Y4cuRIK1OmjGnE7E2VN2rUyNmp9+8Fgfr168feLDQa279/v9tGwV1J+6xatcq0nZfyy1/rVd7mzZtt3LhxbnOdsG3btpkX4FV237593Trv18aNG02zAqkkjdxV97wuS6SSbxD7ejMofpXVtGlTN+syZcoUK126tD388MMp++ZVN7Wv8uXLpyXvvMrzc5naut7wvL8NP/NOd1609XQLn5l/lNu63sv1PhyFtp7KTO55cQ3+22+/tVmzZtn9999vAwcOdFOoXs9NgTkxJS6rUKGCacR3/Phxt+mGDRusevXqpu2UT375e+v1h6DtBwwY4Ebxbdq0ca8Ty+V1+gTUW9csyocffuguy1x11VXpK4ycEUAAgRAInBcjeI2Qa9WqZS+//LLrsel14pT4uc5Ft27d3NS8en1VqlSx5s2bu5H5+vXr7dNPP80zf11r13pde+/evbsbwesavqbjOnTocK4iWZ8GgQsuuCANuZIlAgggED6BrO9GoJl9l0GcuW6aS/WGM43kFaS9pIAvQgXtvPKPX699Evf38kl81BR9qtNHumFw9+7dTNEn4qbxtTdtqfs4opaiPEVPWw++tUW5rUdtir5atWqFOsHnxQjek0k1uCuf+OCu1/HT+XnlH78+r/21jIQAAggggIDfAmdegPa7BPJDAAEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAHzg5BSKAAAIIIJB+AQJ8+o0pAQEEEEAAgcAFCPCBk1MgAggggAAC6RcgwKffmBIQQAABBBAIXIAAX0Dyw4cP229/+1tbs2ZNAfdgMwQQQAABBIpOIGMD/JEjR85QPXr0qJ06deqM5YkLtM3p06dzLf7kk0+sadOmNmfOnFzL9eLkyZPuJ3HFsWPHEhflud0ZG/mw4MMPP7SBAwfaU089ZYcOHfIhR7JAAAEEEIiSQE6UKluQuiqYvfDCC1ahQgXbs2eP3X777VayZEkbP368lS5d2rZu3Wo9e/a0vXv3WlZWlnXp0sU9f+WVV+znP/+5zZw509atW2cnTpywDh06WJs2bVyxCph33XWX/fnPf7b9+/e7/D/++GNbtGiR6zQovxtuuMFq165tY8aMsVKlSpk6GRdddJFdd911tnz5cpd3iRIlrE6dOvbTn/40djjqUCjPxJSdnZ24qECvv/76axswYIB5HQwd89NPP12gfcOyUbFixaywx1+Ux6B6q11Fte5FaZdK2TKPcpuJanuJcltX3TM9ZVyA37Jliwvoffv2jQXNffv2Wffu3a1Zs2a2dOlSW7JkifXu3dtGjx7tAryCdLt27dxId+XKlTZ48GAXtEeMGOECvAKkAnalSpXcdvPnz7drrrkmNsrX9mvXrrW5c+dar169bPv27fbkk0+6tqNpfQX46dOnu3yLFy/uytU2F1xwgdtGnYPhw4fnamuPPPKI1ahRI9eygr5YuHBhLLhrH3Uu1KkgBSOgNw68g7H2SpF59erVvZc8BiRAW08/9MGDBwtdSMYF+AsvvNCaNGlizz//vBux9+nTx8qVK2fvv/++C7IKsOrply9f3v3s2LHDBf0hQ4bYF198Ybt377ZRo0Y5UPWqNbJesGCBmw1QnprmP3DggHXr1s1t06hRI/dYtmxZN+rXi/r168dGcDk5ObGOhspW0j6rVq2KBfiqVavaH/7wB7fO+7Vx40bTSLwwqUGDBm6GwZsV6Ny5c6HzKkz5fuzjzYD4kVeQeZQpU8a1K3XgopbUVnVpSpecopY0c6a/Xf19Ri3R1oM/Y3ov1sxpFNq64ldhU8YF+G3btrmR77Bhw+yjjz6y2bNnm05mq1at7LLLLrPFixe7EbzANAU/ZcoUq1mzphv1K/BWrFjRBg0a5E7+e++950buy5YtM+WnqX4ldQBWrFjhnquzkJgSl+lygabLjx8/7uqyYcMG69ixY+Juvr2uVq2avfXWW+6SgJ5rBoGEAAIIIHB+CWRcgNc03Ztvvmm6KU7X4K+//noXWKdOnerugI8fUbds2dImTJhgd955pzvrCsQtWrRw1/DVu9NNdZqy16yAF9y14Q9/+EM3Hd+2bdsCtxaN+HVJQPlWqVLFmjdvXuB9C7OhRvGPPvpoZEc1hTlm9kEAAQQQ+JdA1ndTcrlvF//Xukg/01SdbmjzbqRQYNWNc1p2ruRN26TjxheN5AtSB03Re/U4V33zW8+0ZX4y6VvOFH36bM+WM239bDrpWRflth61KXrNxBYmZdwI3kOIH3FrmabNCxJYtW06ArvyVSpoHf5va34jgAACCCBQOIEzLyAXLh/2QgABBBBAAIEQCRDgQ3QyqAoCCCCAAAJ+CRDg/ZIkHwQQQAABBEIkQIAP0cmgKggggAACCPglQID3S5J8EEAAAQQQCJEAAT5EJ4OqIIAAAggg4JcAAd4vSfJBAAEEEEAgRAIE+BCdDKqCAAIIIICAXwIEeL8kyQcBBBBAAIEQCRDgQ3QyqAoCCCCAAAJ+CRDg/ZIkHwQQQAABBEIkQIAP0cmgKggggAACCPglQID3S5J8EEAAAQQQCJEAAT5EJ4OqIIAAAggg4JcAAd4vSfJBAAEEEEAgRAIE+BCdDKqCAAIIIICAXwIEeL8kyQcBBBBAAIEQCRDgQ3QyqAoCCCCAAAJ+CRDg/ZIkHwQQQAABBEIkQIAP0cmgKggggAACCPglQID3S5J8EEAAAQQQCJFA1unvUojqQ1V8FPjDH/5gN954ozVq1MjHXMnqbALLly+3+fPn26BBg862Get8FnjmmWesT58+1rBhQ59zJrv8BJYuXWofffSR3XvvvfltwvIiFmAEX8QnIJ3FHz161E6ePJnOIsg7QUDex44dS1jKy3QL0NbTLXxm/qdOnaKtn8kSqiUE+FCdDiqDAAIIIICAPwIEeH8cyQUBBBBAAIFQCXANPlSnw9/KrFmzxmrXrm1ly5b1N2Nyy1dg3759tmfPHu57yFcoPSvU1uvUqWNlypRJTwHkeoYAbf0MktAtIMCH7pRQIQQQQAABBFIXYIo+dUNyQAABBBBAIHQCBPjQnRIqhAACCCCAQOoCOalnQQ5hFJg5c6bpf7IrVKhgN998s5UqVSqM1cyoOn399dc2Y8YM079stW7d2n74wx9m1PGF+WD2799vo0ePtoceeijM1cyYuh04cMDmzJljavMtW7a0zp07Z8yxZdKBMILPpLP5/4/lyy+/tPXr19vgwYPdzV7Tpk3LwKMM3yG9/vrr1qtXL7vnnnvs448/djfbha+WmVkj2evmRlIwAn/961+tRYsW7kNu9F5z8ODBYAqmlKQECPBJcUVj43Xr1tn3v/99y87Otnbt2tnq1aujUfGI13LgwIFWuXJlK1asmGVlZZlGlaT0C2gk2axZMytdunT6C6ME0wfcbN++3YoXL26ffPKJ9e/f38qVK4dMCAUI8CE8KalW6Ztvvon9u5De9L799ttUs2T/AggouCtNnz7dBRs+NrUAaCluokCjf5Hr2LFjijmxe0EFND3vTdFr5K6PxOYTMwuqF+x2BPhgvQMpTUFd14GVjh8/buXLlw+kXAoxe+edd+yrr74yjeZJ6ReYMGGCXXDBBTZ79mw7dOiQLVy4MP2FnucllChRwvQVJv369bOuXbta48aNbdWqVee5SjgPnwAfzvOSUq3q1q1rGzZscHnoerw+7IaUfgEFd41oFNxzcrh/Nf3iZldeeaUL8PqAG10W4WbS9KtrAKGbd48cOeIK0wfe4J5+98KUwAfdFEYt5PvoGtmkSZNMU/W6DqxvNtMfJCl9AnqT+81vfmM1a9Z0gUYlXX/99e5GpPSVSs7xAo8//rg99thj8Yt4niaBtWvX2oIFC1yHtlKlSu4/ddTBIoVLgAAfrvPha230rWaaTiMhgAAC6RDgPSYdqv7lSYD3z5KcEEAAAQQQCI0A1+BDcyqoCALRFtANnfwverTPIbXPLAECfGadT44GgcAFtm3bZtddd51Vr17dLrnkEncfwgsvvBB4PRIL1P0Q+kwIEgLnqwC3+p6vZ57jRsAngaFDh7p/lXrzzTfdfw/og5XatGljP/jBD9wHLflUDNkggECSAozgkwRjcwQQyC2gD1LSHdTevwbqU+X06XL6d02lnTt32o033mi621qfsKj/WfeSnutDavSvnPfff3/sX69mzZrl9qlSpYr7bwTNEig9/fTT7oNVunTp4vK75ZZb7PDhw26dPhf92muvtWrVqtkDDzzgPnFNK06cOGF33nmn275Bgwb2+9//3m3PLwQyXYAAn+lnmONDIM0CjzzyiP3lL3+xiy++2B599FGbN2+eG71rilxJwbVixYr2xRdf2C9+8Qu744473HIF7d69e9vdd9/tvhhp48aN9uc//9n02Q09e/Z0P8uWLXOfCjhgwAC3jzoLTz75pA0bNsyWLl1qixcvtokTJ7p1t912mzVt2tQt1/9oa1slfU69/q1L0/VTp0613/3ud+61W8kvBDJYgLvoM/jkcmgIBCWwdetWe/XVV23y5Mnu0+S6devmAq9uvNO1+c8++8zq1KnjqqPr9SNGjHAB/7e//a0L7lqxadMm27Vrl+mbEJWPOgpKCs4K3PpY2meeecZ9voN3jV9fqFSjRg0bMmSIVa1a1dRp0Cfbbd682c0gaF91BPQtc88995ypXkolS5Z0j/xCIJMFGMFn8tnl2BAIQEAf8lOrVi0XRPXhJxopK+AroCpoa/penzjXvHlz96OgO3/+fPfxppdeemmshprS17V7fdRvhw4dYsubNGnigveWLVvcMgV0L5UtW9ZNwStP7a/grqTOhNeh0AcO6QtR7rrrLrdeMw7eRzl7+fCIQCYKEOAz8axyTAgEJKBAqevnmlb3UsOGDe3WW291I3R9paim55cvX+6CvgK/pt01ba/r6/GfYa4p+jfeeMNdQ1+5cqWXndtP/37XqFEjtyyvT0yrX7++qQOgT29U0n0BO3bscM9VR43gtV6zDFOmTLFx48a5dfxCIJMFCPCZfHY5NgTSLKCp7muuucaNjvW94Er6drfx48e7Ubs+SfGqq65yU/L6CGVNobds2dIF/x/96Efu60Y///xzt59uoNN0evfu3d1NeitWrHA3yo0ePdq+973vuY6C2zCPX7rer86E7gVQeu2119wXLXnP+/bt62YSdBOeZhJICJwPAvyb3PlwljlGBNIooKCqL9hp1aqV+5YxfU+4rolrmZJuiNM3j40dO9ays7Pt4YcfttatW7t1w4cPt/bt27spfk3Fv/zyy246Xp/rr+l7XVfXDIBG9udKutmuV69e7i573S2vUb2Sbr77+9//7v6VT19rqnpqyp6EQKYLcJNdpp9hjg+BgAQ0QteNcPFfuBNftO5q17+wJU6x69/YNKWuQB6ftFzX9xXkk0m7d+/Ocx/dWa/PTueLl5LRZNsoCxDgo3z2qDsCCCCAAAL5CHANPh8YFiOAAAIIIBBlAQJ8lM8edUcAAQQQQCAfAQJ8PjAsRgABBBBAIMoCBPgonz3qjgACCCCAQD4CBPh8YFiMAAIIIIBAlAUI8FE+e9QdAQQQQACBfAQI8PnAsBgBBBBAAIEoCxDgo3z2qDsCCCCAAAL5CPw/2sxI7Upq9CQAAAAASUVORK5CYII=" alt="plot of chunk exploreTime" /> </a>
</div>
</div>
</div>
<h3>Merge results</h3>
<p>Once you have analyzed each chromosome using <code>analyzeChr()</code>, you can use <code>mergeResults()</code> to merge the results. This function does not return an object in R but instead creates several Rdata files with the main results from the different chromosomes.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Go back to the original directory
setwd(originalWd)

## Merge results from several chromosomes. In this case we only have one.
mergeResults(chrs='chr21', prefix="analysisResults",
    genomicState = genomicState$fullGenome, 
    optionsStats = results$optionsStats)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:26 mergeResults: Saving options used
## 2014-09-30 14:46:26 Loading chromosome chr21
## 2014-09-30 14:46:27 mergeResults: calculating FWER
## 2014-09-30 14:46:27 mergeResults: Saving fullNullSummary
## 2014-09-30 14:46:27 mergeResults: Re-calculating the p-values
## 2014-09-30 14:46:27 mergeResults: Saving fullRegions
## 2014-09-30 14:46:27 mergeResults: assigning genomic states
## 2014-09-30 14:46:27 annotateRegions: counting
## 2014-09-30 14:46:27 annotateRegions: annotating
## 2014-09-30 14:46:27 mergeResults: Saving fullAnnotatedRegions
## 2014-09-30 14:46:27 mergeResults: Saving fullFstats
## 2014-09-30 14:46:27 mergeResults: Saving fullTime
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Files created by mergeResults()
dir('analysisResults', pattern = '.Rdata')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "fullAnnotatedRegions.Rdata" "fullFstats.Rdata"          
## [3] "fullNullSummary.Rdata"      "fullRegions.Rdata"         
## [5] "fullTime.Rdata"             "optionsMerge.Rdata"
</code></pre>
</div>
<ul>
<li><em>fullFstats.Rdata</em> contains a list with one element per chromosome. Per chromosome it has the F-statistics.</li>
<li><em>fullNullSummary.Rdata</em> is a list with the summary information from the null regions stored for each chromosome.</li>
<li><em>fullTime.Rdata</em> has the timing information for each chromosome as a list.</li>
</ul>
<h4>optionsMerge</h4>
<p>For reproducibility purposes, the options used the merge the results are stored in <code>optionsMerge</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Options used to merge
load(file.path('analysisResults', 'optionsMerge.Rdata'))

## Contents
names(optionsMerge)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "chrs"            "significantCut"  "minoverlap"      "mergeCall"      
## [5] "cutoffFstatUsed" "optionsStats"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Merge call
optionsMerge$mergeCall</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## mergeResults(chrs = "chr21", prefix = "analysisResults", genomicState = genomicState$fullGenome, 
##     optionsStats = results$optionsStats)
</code></pre>
</div>
<h4>fullRegions</h4>
<p>The main result from <code>mergeResults()</code> is in <code>fullRegions</code>. This is a <code>GRanges</code> object with the candidate DERs from all the chromosomes. It also includes the nearest annotation metadata as well as FWER adjusted p-values (<em>fwer</em>) and whether the FWER adjusted p-value is less than 0.05 (<em>significantFWER</em>).</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Load all the regions
load(file.path('analysisResults', 'fullRegions.Rdata'))

## Metadata columns
names(mcols(fullRegions))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  [1] "value"                      "area"                      
##  [3] "indexStart"                 "indexEnd"                  
##  [5] "cluster"                    "clusterL"                  
##  [7] "meanCoverage"               "meanfetal"                 
##  [9] "meanadult"                  "log2FoldChangeadultvsfetal"
## [11] "pvalues"                    "significant"               
## [13] "qvalues"                    "significantQval"           
## [15] "name"                       "annotation"                
## [17] "description"                "region"                    
## [19] "distance"                   "subregion"                 
## [21] "insidedistance"             "exonnumber"                
## [23] "nexons"                     "UTR"                       
## [25] "annoStrand"                 "geneL"                     
## [27] "codingL"                    "fwer"                      
## [29] "significantFWER"
</code></pre>
</div>
<p>Note that <code>analyzeChr()</code> only has the information for a given chromosome at a time, so <code>mergeResults()</code> re-calculates the p-values and q-values using the information from all the chromosomes.</p>
<h4>fullAnnotatedRegions</h4>
<p>In preparation for visually exploring the results, <code>mergeResults()</code> will run <code>annotateRegions()</code> which counts how many known exons, introns and intragenic segments each candidate DER overlaps (by default with a minimum overlap of 20bp). <code>annotateRegions()</code> uses a summarized version of the genome annotation created with <code>makeGenomicState()</code>. For this example, we can use the data included in <code>derfinder</code> which is the summarized annotation of hg19 for chromosome 21.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Load annotateRegions() output
load(file.path('analysisResults', 'fullAnnotatedRegions.Rdata'))

## Information stored
names(fullAnnotatedRegions)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "countTable"     "annotationList"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Take a peak
lapply(fullAnnotatedRegions, head)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $countTable
##   exon intragenic intron
## 1    1          0      0
## 2    1          0      1
## 3    1          0      0
## 4    1          0      0
## 5    1          0      0
## 6    1          0      0
## 
## $annotationList
## GRangesList object of length 6:
## $1 
## GRanges object with 1 range and 4 metadata columns:
##        seqnames               ranges strand |   theRegion
##           <Rle>            <IRanges>  <Rle> | <character>
##   4625    chr21 [45750346, 45751897]      - |        exon
##                        tx_id                              tx_name
##                <IntegerList>                      <CharacterList>
##   4625 73383,73384,73385,... uc002zeo.2,uc002zep.2,uc002zeq.2,...
##                 gene
##        <IntegerList>
##   4625           263
## 
## $2 
## GRanges object with 2 ranges and 4 metadata columns:
##        seqnames               ranges strand | theRegion
##    753    chr21 [35209291, 35210802]      + |      exon
##   3716    chr21 [35014161, 35275756]      - |    intron
##                        tx_id                              tx_name
##    753 72660,72661,72662,... uc002ysw.3,uc002ysx.3,uc002ysy.3,...
##   3716 73184,73185,73186,... uc002ysj.4,uc002ysk.4,uc002ysm.4,...
##               gene
##    753         235
##   3716 107,189,294
## 
## $3 
## GRanges object with 1 range and 4 metadata columns:
##        seqnames               ranges strand | theRegion             tx_id
##   4698    chr21 [46188495, 46191404]      - |      exon 73406,73407,73408
##                                 tx_name gene
##   4698 uc002zfx.3,uc002zfy.3,uc021wjt.1  260
## 
## ...
## <3 more elements>
## -------
## seqinfo: 1 sequence from hg19 genome
</code></pre>
</div>
<h3>Visually explore results</h3>
<p>Optionally, we can use the addon package <a href="https://github.com/lcolladotor/derfinderPlot">derfinderPlot</a> to visually explore the results. For more details, please check its <a href="lcolladotor.github.io/derfinderPlot/">vignette</a>.</p>
<p>To make the region level plots, we will need to extract the region level coverage data. We can do so using <code>getRegionCoverage()</code> as shown below.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Find overlaps between regions and summarized genomic annotation
annoRegs <- annotateRegions(fullRegions, genomicState$fullGenome)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:28 annotateRegions: counting
## 2014-09-30 14:46:28 annotateRegions: annotating
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Indeed, the result is the same because we only used chr21
identical(annoRegs, fullAnnotatedRegions)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] TRUE
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Get the region coverage
regionCov <- getRegionCoverage(fullCov, fullRegions)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:28 getRegionCoverage: processing chr21
## 2014-09-30 14:46:28 getRegionCoverage: done processing chr21
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore the result
head(regionCov[[1]])</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##   HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178
## 1   0.80   0.80   1.51   1.50   2.05   1.55   1.42   1.54   1.26   1.01
## 2   0.80   0.80   1.51   1.50   2.08   1.55   1.42   1.50   1.26   1.01
## 3   0.80   0.84   1.57   1.50   2.12   1.55   1.46   1.54   1.31   1.01
## 4   0.92   0.84   1.57   1.47   2.12   1.55   1.46   1.54   1.31   1.01
## 5   0.92   0.84   1.57   1.54   2.20   1.55   1.46   1.50   1.31   1.06
## 6   0.96   0.80   1.57   1.50   2.20   1.55   1.42   1.42   1.31   1.10
##   HSB92 HSB97
## 1  1.59  1.27
## 2  1.63  1.27
## 3  1.69  1.37
## 4  1.69  1.37
## 5  1.73  1.41
## 6  1.73  1.37
</code></pre>
</div>
<p>With this, we are all set to visually explore the results.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">library('derfinderPlot')

## Overview of the candidate DERs in the genome
plotOverview(regions = fullRegions, annotation = results$annotation, type = 'fwer')

suppressPackageStartupMessages(library('TxDb.Hsapiens.UCSC.hg19.knownGene'))
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

## Base-levle coverage plots for the first 10 regions
plotRegionCoverage(regions = fullRegions, regionCoverage = regionCov, 
    groupInfo = pheno$group, nearestAnnotation = results$annotation, 
    annotatedRegions = annoRegs, whichRegions=1:10, txdb = txdb, scalefac = 1, 
    ask = FALSE)

## Cluster plot for the first region
plotCluster(idx = 1, regions = fullRegions, annotation = results$annotation, coverageInfo = fullCov$chr21, txdb = txdb, groupInfo = pheno$group, titleUse = 'fwer')</code></pre>
</div>
<h3>Results HTML report</h3>
<p>We have also developed an addon package called <code>regionReport</code> available via <a href="https://github.com/lcolladotor/regionReport">GitHub</a>. For more information check its <a href="http://lcolladotor.github.io/regionReport/">vignette</a>.</p>
<p>The function <code>derfinderReport()</code> in <code>regionReport</code> basically takes advantage of the results from <code>mergeResults()</code> and plotting functions available in <code>derfinderPlot</code> as well as other neat features from <a href="http://rcharts.io/">rCharts</a> and <code>knitrBoostrap</code> <a id='cite-knitrBootstrap'></a>(<a href='https://github.com/jimhester/'>Hester, 2014</a>).</p>
<p>The resulting HTML report promotes reproducibility of the analysis and allows you to explore in more detail the results through some diagnostic plots.</p>
<h2>Region matrix analysis</h2>
<p>An alternative type of analysis is driven by <code>regionMatrix()</code>. The idea is to consider consecutive bases with mean coverage above a given cutoff as potentially differentially expressed regions. Then, get a matrix where each row is one of these regions and each column represents a sample. Each cell of the matrix has the mean coverage for the specific region - sample pair. Then, other packages specialized in differential expression at the counts level can be used, for example <a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html">limma</a>.</p>
<p>In this example, we will use <code>regionMatrix()</code> where we filter by mean coverage greater than 30 out of counts standardized to libraries of 40 million reads. Note that read of the <em>BrainSpan</em> data are 76bp long.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Use regionMatrix()
system.time(regionMat <- regionMatrix(fullCov, cutoff = 30, L = 76, totalMapped = 2^(sampleDepths), targetSize = 40e6))</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:29 regionMatrix: processing chr21
## 2014-09-30 14:46:29 filterData: normalizing coverage
## 2014-09-30 14:46:29 filterData: done normalizing coverage
## 2014-09-30 14:46:30 filterData: originally there were 48129895 rows, now there are 189651 rows. Meaning that 99.61 percent was filtered.
## 2014-09-30 14:46:30 findRegions: identifying potential segments
## 2014-09-30 14:46:30 findRegions: segmenting F-stats information
## 2014-09-30 14:46:31 findRegions: identifying candidate DERs
## 2014-09-30 14:46:31 findRegions: identifying DER clusters
## 2014-09-30 14:46:34 getRegionCoverage: processing chr21
## 2014-09-30 14:46:36 getRegionCoverage: done processing chr21
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##   7.137   0.823   7.926
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore results
names(regionMat$chr21)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "regions"        "bpCoverage"     "coverageMatrix"
</code></pre>
</div>
<p><code>regionMatrix()</code> returns three pieces of output.</p>
<ul>
<li><em>regions</em> is the result from filtering with <code>filterData()</code> and then defining the regions with <code>findRegions()</code>. Note that the metadata variable <code>value</code> represents the mean coverage for the given region while <code>area</code> is the sum of the base-level coverage (before adjusting for read length) from all samples.</li>
<li><em>bpCoverage</em> is a list with the base-level coverage from all the regions. This information can then be used with <code>plotRegionCoverage()</code> from <code>derfinderPlot</code>.</li>
<li><em>coverageMatrix</em> is the matrix with one row per region and one column per sample. The different matrices for each of the chromosomes can then be merged prior to using <a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html">limma</a> or other packages. Note that the counts are generally not integers, but can easily be rounded if necessary.</li>
</ul>
<p>Similar to what we saw earlier, the regions are arranged in a <code>GRanges</code> object. In this case, the metadata is simpler because no annotation information is included.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## regions output
regionMat$chr21$regions</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## GRanges object with 1822 ranges and 6 metadata columns:
##        seqnames               ranges strand   |            value
##           <Rle>            <IRanges>  <Rle>   |        <numeric>
##      1    chr21   [9825461, 9825583]      *   | 64.1835795934005
##      2    chr21   [9825645, 9825646]      *   | 30.6206716563444
##      3    chr21   [9825872, 9826023]      *   | 41.4259978624777
##      4    chr21   [9826149, 9826225]      *   | 34.5680066935166
##      5    chr21   [9826990, 9827584]      *   | 12578.8649688978
##    ...      ...                  ...    ... ...              ...
##   1818    chr21 [48084207, 48084835]      *   | 227.754828611473
##   1819    chr21 [48085444, 48085455]      *   | 31.5989673322147
##   1820    chr21 [48085458, 48085458]      *   |  30.078238709839
##   1821    chr21 [48085469, 48085470]      *   | 30.3090276995855
##   1822    chr21 [48085475, 48085477]      *   | 30.5711564090198
##                    area indexStart  indexEnd cluster clusterL
##               <numeric>  <integer> <integer>   <Rle>    <Rle>
##      1 7894.58028998826          1       123       1      765
##      2 61.2413433126887        124       125       1      765
##      3 6296.75167509661        126       277       1      765
##      4 2661.73651540077        278       354       1      765
##      5 7484424.65649418        355       949       2      595
##    ...              ...        ...       ...     ...      ...
##   1818 143257.787196617     189005    189633     800      629
##   1819 379.187607986577     189634    189645     801       34
##   1820  30.078238709839     189646    189646     801       34
##   1821  60.618055399171     189647    189648     801       34
##   1822 91.7134692270595     189649    189651     801       34
##   -------
##   seqinfo: 1 sequence from an unspecified genome
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Number of regions
length(regionMat$chr21$regions)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] 1822
</code></pre>
</div>
<p><code>bpCoverage</code> is the base level coverage list which can then be used for plotting.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Base-level coverage matrices for each of the regions
## Useful for plotting
lapply(regionMat$chr21$bpCoverage[1:2], head, n = 2)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $`1`
##   HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178
## 1  117.9  4.033  3.169  3.724  8.172  6.974      0  69.91  22.58   7.86
## 2  117.9  4.033  3.169  5.852  8.172  6.974      0  69.91  22.58   7.86
##   HSB92 HSB97
## 1 6.732 110.7
## 2 6.732 110.7
## 
## $`2`
##     HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178
## 124  130.4  8.643  12.68   13.3   9.08  17.05  9.182   36.1  38.71  10.67
## 125  128.7  8.643  12.68   13.3   9.08  17.05  9.182   36.1  38.71  10.67
##     HSB92 HSB97
## 124 13.98 67.41
## 125 13.98 69.60
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Check dimensions. First region is 123 long, second one is 2 bp long.
## The columns match the number of samples (12 in this case).
lapply(regionMat$chr21$bpCoverage[1:2], dim)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## $`1`
## [1] 123  12
## 
## $`2`
## [1]  2 12
</code></pre>
</div>
<p>The end result of the coverage matrix is shown below. Note that the coverage has been adjusted for read length. Because reads might not fully align inside a given region, the numbers are generally not integers but can be rounded if needed.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Dimensions of the coverage matrix
dim(regionMat$chr21$coverageMatrix)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] 1822   12
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Coverage for each region. This matrix can then be used with limma or other pkgs
head(regionMat$chr21$coverageMatrix)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##      HSB113    HSB123    HSB126    HSB130    HSB135    HSB136    HSB145
## 1 4.464e+02 1.490e+01 1.647e+01    34.761 3.460e+01 4.345e+01 4.733e+00
## 2 3.409e+00 2.274e-01 3.336e-01     0.350 2.389e-01 4.486e-01 2.416e-01
## 3 1.674e+02 1.058e+01 1.822e+01    48.201 1.784e+01 3.630e+01 1.333e+01
## 4 1.524e+02 7.536e+00 7.256e+00     9.891 8.454e+00 3.898e+01 1.064e+01
## 5 1.578e+05 1.597e+04 7.381e+04 58888.964 2.721e+05 2.159e+05 7.186e+04
## 6 1.450e+04 5.775e+03 2.457e+04 14295.996 3.643e+04 4.520e+04 1.394e+04
##      HSB153    HSB159    HSB178     HSB92     HSB97
## 1 2.160e+02    83.678 1.806e+01 4.329e+01   290.092
## 2 9.501e-01     1.019 2.807e-01 3.679e-01     1.803
## 3 4.148e+01    18.973 0.000e+00 1.211e+01   609.755
## 4 8.007e+01     5.560 0.000e+00 7.393e+00    92.053
## 5 8.379e+04 19258.506 7.904e+04 6.055e+04 72717.001
## 6 1.806e+04  4400.059 1.092e+04 6.653e+03 11113.935
</code></pre>
</div>
<h2>Feature level analysis</h2>
<p>Similar to the region level analysis, you might be interested in performing a feature level analysis. More specifically, this could be getting a count matrix at the exon level. <code>coverageToExon()</code> allows you to get such a matrix by taking advantage of the summarized annotation produced by <code>makeGenomicState()</code>.</p>
<p>In this example, we use the genomic state included in the package which has the information for chr21 <code>TxDb.Hsapiens.UCSC.hg19.knownGene</code> annotation.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Get the exon level matrix
system.time(exonCov <- coverageToExon(fullCov, genomicState$fullGenome, L = 76))</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:41 coverageToExon: processing chromosome chr21
## 2014-09-30 14:46:46 coverageToExon: processing chromosome chr21
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##   8.782   2.871  11.612
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Dimensions of the matrix
dim(exonCov)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] 2658   12
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Explore a little bit
tail(exonCov)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##      HSB113   HSB123   HSB126   HSB130    HSB135    HSB136    HSB145
## 4983 0.1174   0.1383   0.0000   0.0000   0.07895   0.03947 5.263e-03
## 4985 4.6542   1.4558   1.6034   0.7799   1.29592   0.72987 9.964e-01
## 4986 7.1511 291.0205 252.4892 152.1905 439.69500 263.63487 2.345e+02
## 4988 0.0000   0.7063   0.7224   0.2639   0.48658   0.37658 6.157e-01
## 4990 2.3064  64.9424  69.6584  35.5770 101.76395  76.96737 5.979e+01
## 4992 0.1653   8.1834   9.7538   2.4193  10.08618  14.41013 5.326e+00
##       HSB153   HSB159 HSB178    HSB92    HSB97
## 4983  0.1053  0.04934 0.1289  0.03987  0.44605
## 4985 10.4907  4.39013 7.3120  1.65184 10.87724
## 4986  4.5311 10.64368 4.9808 11.40447  6.23316
## 4988  0.0000  0.00000 0.0000  0.05645  0.00000
## 4990  1.2284  2.67487 0.9542  3.98013  1.96132
## 4992  0.2403  0.73039 0.2601  0.66908  0.07474
</code></pre>
</div>
<p>With this matrix, rounded if necessary, you can proceed to use packages such as <a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html">limma</a>, <a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq</a>, <a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>, among others.</p>
<h2>Compare results visually</h2>
<p>We can certainly make region level plots using <code>plotRegionCoverage()</code> or cluster plots using <code>plotCluster()</code> or overview plots using <code>plotOveview()</code>, all from <code>derfinderPlot</code>.</p>
<p>First we need to get the relevant annotation information.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Annotate regions as exonic, intronic or intragenic
system.time(annoGenome <- annotateRegions(regionMat$chr21$regions, genomicState$fullGenome))</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:46:49 annotateRegions: counting
## 2014-09-30 14:46:50 annotateRegions: annotating
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##   0.432   0.004   0.450
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Note that the genomicState object included in derfinder only has information
## for chr21 (hg19).

## Identify closest genes to regions
suppressPackageStartupMessages(library('bumphunter'))
system.time(annoNear <- annotateNearest(regionMat$chr21$regions, subject = 'hg19'))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## Matching regions to genes.
## nearestgene: loading bumphunter hg19 transcript database
## finding nearest transcripts...
## Annotating.Done.
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##    user  system elapsed 
##  81.383   0.399  81.474
</code></pre>
</div>
<p>Now we can proceed to use <code>derfinderPlot</code> to make the region level plots for the top 100 regions.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Identify the top regions by highest total coverage
top <- order(regionMat$chr21$regions$area, decreasing = TRUE)[1:100]

## Base-level plots for the top 100 regions with transcript information
suppressPackageStartupMessages(library('TxDb.Hsapiens.UCSC.hg19.knownGene'))
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

library('derfinderPlot')
plotRegionCoverage(regionMat$chr21$regions, regionCoverage = regionMat$chr21$bpCoverage, groupInfo = pheno$group, nearestAnnotation = annoNear, annotatedRegions = annoGenome, whichRegions = top, scalefac = 1, txdb = txdb, ask = FALSE)</code></pre>
</div>
<p>However, we can alternatively use <a href="http://www.bioconductor.org/packages/release/bioc/html/epivizr.html">epivizr</a> to view the candidate DERs and the region matrix results in a genome browser.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Load epivizr, it's available from Bioconductor
library('epivizr')

## Load data to your browser
mgr <- startEpiviz()
ders_dev <- mgr$addDevice(
    fullRegions[as.logical(fullRegions$significantFWER) ], "Candidate DERs")
ders_potential_dev <- mgr$addDevice(
    fullRegions[!as.logical(fullRegions$significantFWER) ], "Potential DERs")
regs_dev <- mgr$addDevice(regionMat$chr21$regions, "Region Matrix")

## Go to a place you like in the genome
mgr$navigate("chr21", start(regionMat$chr21$regions[top[1]]) - 100, end(regionMat$chr21$regions[top[1]]) + 100)

## Stop the navigation
mgr$stopServer()</code></pre>
</div>
<h2>Export coverage to BigWig files</h2>
<p><code>derfinder</code> also includes <code>createBw()</code> with related functions <code>createBwSample()</code> and <code>coerceGR()</code> to export the output of <code>fullCoverage()</code> to BigWig files. These functions can be useful in the case where you start with BAM files and later on want to save the coverage data into BigWig files, which are generally smaller.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Subset only the first sample
fullCovSmall <- lapply(fullCov, '[', 1)

## Export to BigWig
bw <- createBw(fullCovSmall)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-down"></span> R message
</button>
<pre style=""><code class="message r">## 2014-09-30 14:48:12 coerceGR: coercing sample HSB113
## 2014-09-30 14:48:12 createBwSample: exporting bw for sample HSB113
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## See the file. Note that the sample name is used to name the file.
dir(pattern = '.bw')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "HSB113.bw"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Internally createBw() coerces each sample to a GRanges object before 
## exporting to a BigWig file. If more than one sample was exported, the
## GRangesList would have more elements.
bw</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## GRangesList object of length 1:
## $HSB113 
## GRanges object with 155950 ranges and 1 metadata column:
##         seqnames               ranges strand   |              score
##            <Rle>            <IRanges>  <Rle>   |          <numeric>
##   chr21    chr21   [9458667, 9458741]      *   | 0.0399999991059303
##   chr21    chr21   [9540957, 9540971]      *   | 0.0399999991059303
##   chr21    chr21   [9543719, 9543778]      *   | 0.0399999991059303
##   chr21    chr21   [9651480, 9651554]      *   | 0.0399999991059303
##   chr21    chr21   [9653397, 9653471]      *   | 0.0399999991059303
##     ...      ...                  ...    ... ...                ...
##   chr21    chr21 [48093246, 48093255]      *   | 0.0399999991059303
##   chr21    chr21 [48093257, 48093331]      *   | 0.0399999991059303
##   chr21    chr21 [48093350, 48093424]      *   | 0.0399999991059303
##   chr21    chr21 [48112194, 48112268]      *   | 0.0399999991059303
##   chr21    chr21 [48115056, 48115130]      *   | 0.0399999991059303
## 
## -------
## seqinfo: 1 sequence from an unspecified genome
</code></pre>
</div>
<h1>Advanced arguments</h1>
<p>If you are interested in using the advanced arguments, use <code>advancedArg()</code> as shown below:</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## URLs to advanced arguemtns
sapply(c('analyzeChr', 'loadCoverage'), advancedArg, browse = FALSE)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##                                                                                                                             analyzeChr 
## "https://github.com/search?utf8=%E2%9C%93&q=advanced_argument+filename%3Aanalyze+repo%3Alcolladotor%2Fderfinder+path%3A%2FR&type=Code" 
##                                                                                                                           loadCoverage 
##    "https://github.com/search?utf8=%E2%9C%93&q=advanced_argument+filename%3Aload+repo%3Alcolladotor%2Fderfinder+path%3A%2FR&type=Code"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Set browse = TRUE if you want to open them in your browser</code></pre>
</div>
<p>The most common advanced arguments are <code>chrsStyle</code> (default is <code>UCSC</code>) and <code>verbose</code> (by default <code>TRUE</code>). <code>chrsStyle</code> is used to determine the chromosome naming style and is powered by <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html">GenomeInfoDb</a>. <code>verbose</code> controls whether to print status updates for nearly all the functions.</p>
<p>The third commonly used advanced argument is <code>mc.cores</code>. It controls the number of cores to use for the functions that can run with more than one core to speed up. In nearly all the cases, the maximum number of cores depends on the number of chromosomes. One notable exception is <code>analyzeChr()</code> where the maximum number of cores depends on the <code>chunksize</code> used and the dimensions of the data for the chromosome under study.</p>
<h1>Summary</h1>
<p>We have illustrated how to identify candidate differentially expressed regions without using annotation in the identification process by using <code>analyzeChr()</code>. Furthermore, we covered how to perform the region matrix analysis with <code>regionMatrix()</code>. We also highlighted other uses of the <code>derfinder</code> package.</p>
<h1>Origins</h1>
<p>This implementation of <code>derfinder</code> (<a href='https://github.com/lcolladotor/derfinder'>Collado-Torres, Frazee, Jaffe, and Leek, 2014</a>) has its origins in <a href="https://github.com/alyssafrazee/derfinder">Alyssa C. Frazee's derfinder</a> <a id='cite-originalder'></a>(<a href='#bib-originalder'>Frazee, Sabunciyan, Hansen, Irizzary, et al., 2013</a>). The statistical methods and implementation by now are very different.</p>
<h1>Citing <code>derfinder</code></h1>
<p>Please use:</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Citation info
citation('derfinder')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## 
## To cite package 'derfinder' in publications use:
## 
##   Leonardo Collado-Torres, Alyssa Frazee, Andrew Jaffe and Jeffrey
##   Leek (2014). derfinder: Fast differential expression analysis of
##   RNA-seq data at base-pair resolution. R package version 0.99.4.
##   https://github.com/lcolladotor/derfinder
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {derfinder: Fast differential expression analysis of RNA-seq data at base-pair
## resolution},
##     author = {Leonardo Collado-Torres and Alyssa Frazee and Andrew Jaffe and Jeffrey Leek},
##     year = {2014},
##     note = {R package version 0.99.4},
##     url = {https://github.com/lcolladotor/derfinder},
##   }
</code></pre>
</div>
<h1>Reproducibility</h1>
<p>This package was made possible thanks to:</p>
<ul>
<li>R <a id='cite-R'></a>(<a href='http://www.R-project.org/'>R Core Team, 2014</a>)</li>
<li>AnnotationDbi <a id='cite-AnnotationDbi'></a>(<a href='#bib-AnnotationDbi'>Pages, Carlson, Falcon, and Li, 2014</a>)</li>
<li>BiocParallel <a id='cite-BiocParallel'></a>(<a href='#bib-BiocParallel'>Morgan, Lang, and Thompson, 2014</a>)</li>
<li>bumphunter <a id='cite-bumphunter'></a>(<a href='#bib-bumphunter'>Irizarry, Aryee, Bravo, Hansen, et al., 2014</a>) and <a id='cite-bumphunterPaper'></a>(<a href='#bib-bumphunterPaper'>Jaffe, E, Murakami, Peter, et al., 2012</a>)</li>
<li>derfinderHelper <a id='cite-derfinderHelper'></a>(<a href='https://github.com/lcolladotor/derfinderHelper'>Collado-Torres, Jaffe, and Leek, 2014</a>)</li>
<li>GenomeInfoDb <a id='cite-GenomeInfoDb'></a>(<a href='#bib-GenomeInfoDb'>Arora, Morgan, Carlson, and Pages, 2014</a>)</li>
<li>GenomicAlignments <a id='cite-GenomicAlignments'></a>(<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pags, Aboyoun, et al., 2013</a>)</li>
<li>GenomicFeatures (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pags, Aboyoun, et al., 2013</a>)</li>
<li>GenomicFiles <a id='cite-GenomicFiles'></a>(<a href='#bib-GenomicFiles'>Obenchain, Love, and Morgan, 2014</a>)</li>
<li>GenomicRanges (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pags, Aboyoun, et al., 2013</a>)</li>
<li>Hmisc <a id='cite-Hmisc'></a>(<a href='http://CRAN.R-project.org/package=Hmisc'>Jr, with contributions from Charles Dupont, and many others., 2014</a>)</li>
<li>IRanges (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pags, Aboyoun, et al., 2013</a>)</li>
<li>qvalue <a id='cite-qvalue'></a>(<a href='#bib-qvalue'>Dabney and Storey, 2014</a>)</li>
<li>Rsamtools <a id='cite-Rsamtools'></a>(<a href='http://bioconductor.org/packages/release/bioc/html/Rsamtools.html'>Morgan, Pags, Obenchain, and Hayden, 2014</a>)</li>
<li>rtracklayer <a id='cite-rtracklayer'></a>(<a href='http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract'>Lawrence, Gentleman, and Carey, 2009</a>)</li>
<li>S4Vectors <a id='cite-S4Vectors'></a>(<a href='#bib-S4Vectors'>Pages, Lawrence, and Aboyoun, 2014</a>)</li>
<li>biovizBase <a id='cite-biovizBase'></a>(<a href='#bib-biovizBase'>Yin, Lawrence, and Cook, 2014</a>)</li>
<li>derfinderData (<a href='https://github.com/lcolladotor/derfinderData'>Collado-Torres, Jaffe, and Leek, 2014</a>)</li>
<li>devtools <a id='cite-devtools'></a>(<a href='http://CRAN.R-project.org/package=devtools'>Wickham and Chang, 2014</a>)</li>
<li>ggplot2 <a id='cite-ggplot2'></a>(<a href='http://had.co.nz/ggplot2/book'>Wickham, 2009</a>)</li>
<li>knitcitations <a id='cite-knitcitations'></a>(<a href='https://github.com/cboettig/knitcitations'>Boettiger, 2014</a>)</li>
<li>knitr <a id='cite-knitr'></a>(<a href='http://www.crcpress.com/product/isbn/9781466561595'>Xie, 2014</a>)</li>
<li>knitrBootstrap (<a href='https://github.com/jimhester/'>Hester, 2014</a>)</li>
<li>rmarkdown <a id='cite-rmarkdown'></a>(<a href='http://CRAN.R-project.org/package=rmarkdown'>Allaire, McPherson, Xie, Wickham, et al., 2014</a>)</li>
<li>testthat <a id='cite-testthat'></a>(<a href='http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf'>Wickham, 2011</a>)</li>
<li>TxDb.Hsapiens.UCSC.hg19.knownGene <a id='cite-TxDbHsapiensUCSChg19knownGene'></a>(<a href='#bib-TxDbHsapiensUCSChg19knownGene'>Carlson, 2014</a>)</li>
</ul>
<p>Code for creating the vignette</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-up"></span> R source
</button>
<pre style="display:none"><code class="source r">## Create the vignette
library('knitrBootstrap') 

knitrBootstrapFlag <- packageVersion('knitrBootstrap') < '1.0.0'
if(knitrBootstrapFlag) {
    ## CRAN version
    library('knitrBootstrap')
    system.time(knit_bootstrap('derfinder.Rmd', chooser=c('boot', 'code'), show_code = TRUE))
    unlink('derfinder.md')
} else {
    ## GitHub version
    library('rmarkdown')
    system.time(render('derfinder.Rmd', 'knitrBootstrap::bootstrap_document'))
}
## Note: if you prefer the knitr version use:
# library('rmarkdown')
# system.time(render('derfinder.Rmd', 'html_document'))
## Clean up
unlink('analysisResults', recursive = TRUE)
file.remove('HSB97.bw')
file.remove('derfinderRef.bib')

## Extract the R code
library('knitr')
knit('derfinder.Rmd', tangle = TRUE)</code></pre>
</div>
<p>Date the vignette was generated.</p>
<div class="row">
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "2014-09-30 14:48:13 EDT"
</code></pre>
</div>
<p>Wallclock time spent generating the vignette.</p>
<div class="row">
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## Time difference of 2.522 mins
</code></pre>
</div>
<p><code>R</code> session information.</p>
<div class="row">
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## Session info--------------------------------------------------------------
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  setting  value                       
##  version  R version 3.1.1 (2014-07-10)
##  system   x86_64, darwin10.8.0        
##  ui       AQUA                        
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  tz       America/New_York
</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## Packages------------------------------------------------------------------
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  package           * version  date      
##  acepack             1.3.3.3  2013-05-03
##  AnnotationDbi       1.27.16  2014-09-26
##  base64enc           0.1.2    2014-06-26
##  BatchJobs           1.4      2014-09-24
##  BBmisc              1.7      2014-06-21
##  bibtex              0.3.6    2013-07-29
##  Biobase             2.25.0   2014-04-12
##  BiocGenerics      * 0.11.5   2014-09-13
##  BiocParallel        0.99.22  2014-09-24
##  biomaRt             2.21.1   2014-07-08
##  Biostrings          2.33.14  2014-09-09
##  bitops              1.0.6    2013-08-17
##  brew                1.0.6    2011-04-13
##  bumphunter        * 1.5.5    2014-08-29
##  checkmate           1.4      2014-09-03
##  cluster             1.15.3   2014-09-04
##  codetools           0.2.9    2014-08-21
##  colorspace          1.2.4    2013-09-30
##  DBI                 0.3.1    2014-09-24
##  derfinder         * 0.99.4   2014-09-30
##  derfinderData     * 0.99.0   2014-09-30
##  derfinderHelper     0.99.2   2014-09-29
##  devtools          * 1.6      2014-09-23
##  digest              0.6.4    2013-12-03
##  doRNG               1.6      2014-03-07
##  evaluate            0.5.5    2014-04-29
##  fail                1.2      2013-09-19
##  foreach           * 1.4.2    2014-04-11
##  foreign             0.8.61   2014-03-28
##  formatR             1.0      2014-08-25
##  Formula             1.1.2    2014-07-13
##  futile.logger       1.3.7    2014-01-23
##  futile.options      1.0.0    2010-04-06
##  GenomeInfoDb      * 1.1.23   2014-09-28
##  GenomicAlignments   1.1.30   2014-09-23
##  GenomicFeatures     1.17.18  2014-09-29
##  GenomicFiles        1.1.19   2014-08-26
##  GenomicRanges     * 1.17.42  2014-09-23
##  ggplot2           * 1.0.0    2014-05-21
##  gtable              0.1.2    2012-12-05
##  Hmisc               3.14.5   2014-09-12
##  htmltools           0.2.6    2014-09-08
##  httr                0.5      2014-09-02
##  IRanges           * 1.99.28  2014-09-10
##  iterators         * 1.0.7    2014-04-11
##  knitcitations     * 1.0.2    2014-09-30
##  knitr             * 1.6      2014-05-25
##  knitrBootstrap    * 1.0.0    2014-09-30
##  labeling            0.3      2014-08-23
##  lambda.r            1.1.6    2014-01-23
##  lattice             0.20.29  2014-04-04
##  latticeExtra        0.6.26   2013-08-15
##  locfit            * 1.5.9.1  2013-04-20
##  lubridate           1.3.3    2013-12-31
##  markdown            0.7.4    2014-08-24
##  MASS                7.3.34   2014-08-21
##  Matrix              1.1.4    2014-06-15
##  matrixStats         0.10.0   2014-06-08
##  memoise             0.2.1    2014-04-22
##  mime                0.2      2014-09-26
##  munsell             0.4.2    2013-07-11
##  nnet                7.3.8    2014-03-28
##  pkgmaker            0.22     2014-05-14
##  plyr                1.8.1    2014-02-26
##  proto               0.3.10   2012-12-22
##  qvalue              1.39.1   2014-05-27
##  R.methodsS3         1.6.1    2014-01-05
##  RColorBrewer        1.0.5    2011-06-17
##  Rcpp                0.11.2   2014-06-08
##  RCurl               1.95.4.3 2014-07-29
##  RefManageR          0.8.34   2014-08-19
##  registry            0.2      2012-01-24
##  reshape2            1.4      2014-04-23
##  RJSONIO             1.3.0    2014-07-28
##  rmarkdown         * 0.3.3    2014-09-17
##  rngtools            1.2.4    2014-03-06
##  rpart               4.1.8    2014-03-28
##  Rsamtools           1.17.34  2014-09-20
##  RSQLite             0.11.4   2013-05-26
##  rstudioapi          0.1      2014-03-27
##  rtracklayer         1.25.16  2014-09-10
##  S4Vectors         * 0.2.4    2014-09-14
##  scales              0.2.4    2014-04-22
##  sendmailR           1.2.1    2014-09-21
##  stringr             0.6.2    2012-12-06
##  survival            2.37.7   2014-01-22
##  XML                 3.98.1.1 2013-06-20
##  xtable              1.7.4    2014-09-12
##  XVector             0.5.8    2014-09-07
##  yaml                2.1.13   2014-06-12
##  zlibbioc            1.11.1   2014-04-19
##  source                                      
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  Github (lcolladotor/derfinder@031bb0f)      
##  Github (lcolladotor/derfinderData@75a44a6)  
##  Github (lcolladotor/derfinderHelper@54a9721)
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  Github (cboettig/knitcitations@f567713)     
##  CRAN (R 3.1.0)                              
##  Github (jimhester/knitrBootstrap@76c41f0)   
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  Bioconductor
</code></pre>
</div>
<h1>Bibliography</h1>
<p>This vignette was generated using <code>knitrBootstrap</code> (<a href='https://github.com/jimhester/'>Hester, 2014</a>) with <code>knitr</code> (<a href='http://www.crcpress.com/product/isbn/9781466561595'>Xie, 2014</a>) and <code>rmarkdown</code> (<a href='http://CRAN.R-project.org/package=rmarkdown'>Allaire, McPherson, Xie, Wickham, et al., 2014</a>) running behind the scenes.</p>
<p>Citations made with <code>knitcitations</code> (<a href='https://github.com/cboettig/knitcitations'>Boettiger, 2014</a>).</p>
<div class="row">
<p>
<a id='bib-rmarkdown'></a><a href="#cite-rmarkdown">[1]</a><cite> J. Allaire, J. McPherson, Y. Xie, H. Wickham, et al. <EM>rmarkdown: Dynamic Documents for R</EM>. R package version 0.3.3. 2014. URL: <a href="http://CRAN.R-project.org/package=rmarkdown">http://CRAN.R-project.org/package=rmarkdown</a>.</cite>
</p>
<p>
<a id='bib-GenomeInfoDb'></a><a href="#cite-GenomeInfoDb">[2]</a><cite> S. Arora, M. Morgan, M. Carlson and H. Pages. <EM>GenomeInfoDb: Utilities for manipulating chromosome and other 'seqname' identifiers</EM>. R package version 1.1.23. 2014.</cite>
</p>
<p>
<a id='bib-knitcitations'></a><a href="#cite-knitcitations">[3]</a><cite> C. Boettiger. <EM>knitcitations: Citations for knitr markdown files</EM>. R package version 1.0.2. 2014. URL: <a href="https://github.com/cboettig/knitcitations">https://github.com/cboettig/knitcitations</a>.</cite>
</p>
<p>
<a id='bib-brainspan'></a><a href="#cite-brainspan">[4]</a><cite> BrainSpan. Atlas of the Developing Human Brain [Internet]. Funded by ARRA Awards 1RC2MH089921-01, 1RC2MH090047-01, and 1RC2MH089929-01. 2011. URL: <a href="http://developinghumanbrain.org">http://developinghumanbrain.org</a>.</cite>
</p>
<p>
<a id='bib-TxDbHsapiensUCSChg19knownGene'></a><a href="#cite-TxDbHsapiensUCSChg19knownGene">[5]</a><cite> M. Carlson. <EM>TxDb.Hsapiens.UCSC.hg19.knownGene: Annotation package for TxDb object(s)</EM>. R package version 3.0.0. 2014.</cite>
</p>
<p>
<a id='bib-derfinder'></a><a href="#cite-derfinder">[6]</a><cite> L. Collado-Torres, A. Frazee, A. Jaffe and J. Leek. <EM>derfinder: Fast differential expression analysis of RNA-seq data at base-pair resolution</EM>. R package version 0.99.4. 2014. URL: <a href="https://github.com/lcolladotor/derfinder">https://github.com/lcolladotor/derfinder</a>.</cite>
</p>
<p>
<a id='bib-derfinderData'></a><a href="#cite-derfinderData">[7]</a><cite> L. Collado-Torres, A. Jaffe and J. Leek. <EM>derfinderData: Processed BigWigs from BrainSpan for examples</EM>. R package version 0.99.0. 2014. URL: <a href="https://github.com/lcolladotor/derfinderData">https://github.com/lcolladotor/derfinderData</a>.</cite>
</p>
<p>
<a id='bib-derfinderHelper'></a><a href="#cite-derfinderHelper">[8]</a><cite> L. Collado-Torres, A. Jaffe and J. Leek. <EM>derfinderHelper: derfinder helper package</EM>. R package version 0.99.2. 2014. URL: <a href="https://github.com/lcolladotor/derfinderHelper">https://github.com/lcolladotor/derfinderHelper</a>.</cite>
</p>
<p>
<a id='bib-qvalue'></a><a href="#cite-qvalue">[9]</a><cite> A. Dabney and J. D. Storey. <EM>qvalue: Q-value estimation for false discovery rate control</EM>. R package version 1.39.1. 2014.</cite>
</p>
<p>
<a id='bib-originalder'></a><a href="#cite-originalder">[10]</a><cite> A. C. Frazee, S. Sabunciyan, K. D. Hansen, R. A. Irizzary, et al. Differential expression analysis of RNA-seq data at single-base resolution. In: <EM>Biostatistics</EM> (2013).</cite>
</p>
<p>
<a id='bib-knitrBootstrap'></a><a href="#cite-knitrBootstrap">[11]</a><cite> J. Hester. <EM>knitrBootstrap: Knitr Bootstrap framework.</EM> R package version 1.0.0. 2014. URL: <a href="https://github.com/jimhester/">https://github.com/jimhester/</a>.</cite>
</p>
<p>
<a id='bib-bumphunter'></a><a href="#cite-bumphunter">[12]</a><cite> R. A. Irizarry, M. Aryee, H. C. Bravo, K. D. Hansen, et al. <EM>bumphunter: Bump Hunter</EM>. R package version 1.5.5. 2014.</cite>
</p>
<p>
<a id='bib-bumphunterPaper'></a><a href="#cite-bumphunterPaper">[13]</a><cite> Jaffe, A. E, Murakami, Peter, et al. Bump hunting to identify differentially methylated regions in epigenetic epidemiology studies. In: <EM>International Journal of Epidemiology</EM> (2012).</cite>
</p>
<p>
<a id='bib-Hmisc'></a><a href="#cite-Hmisc">[14]</a><cite> F. E. H. Jr, with contributions from Charles Dupont and many others. <EM>Hmisc: Harrell Miscellaneous</EM>. R package version 3.14-5. 2014. URL: <a href="http://CRAN.R-project.org/package=Hmisc">http://CRAN.R-project.org/package=Hmisc</a>.</cite>
</p>
<p>
<a id='bib-rtracklayer'></a><a href="#cite-rtracklayer">[15]</a><cite> M. Lawrence, R. Gentleman and V. Carey. rtracklayer: an R package for interfacing with genome browsers. In: <EM>Bioinformatics</EM> 25 (2009), pp. 1841-1842. DOI: <a href="http://dx.doi.org/10.1093/bioinformatics/btp328">10.1093/bioinformatics/btp328</a>. URL: <a href="http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract">http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract</a>.</cite>
</p>
<p>
<a id='bib-GenomicAlignments'></a><a href="#cite-GenomicAlignments">[16]</a><cite> M. Lawrence, W. Huber, H. Pags, P. Aboyoun, et al. Software for Computing and Annotating Genomic Ranges. In: <EM>PLoS Computational Biology</EM> 9 (8 2013). DOI: <a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>. URL: <a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118}.">http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118}.</a>
</p>
</cite>
</p>
<p>
<a id='bib-BiocParallel'></a><a href="#cite-BiocParallel">[17]</a><cite> M. Morgan, M. Lang and R. Thompson. <EM>BiocParallel: Bioconductor facilities for parallel evaluation</EM>. R package version 0.99.22. 2014.</cite>
</p>
<p>
<a id='bib-Rsamtools'></a><a href="#cite-Rsamtools">[18]</a><cite> M. Morgan, H. Pags, V. Obenchain and N. Hayden. <EM>Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import</EM>. R package version 1.17.34. 2014. URL: <a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">http://bioconductor.org/packages/release/bioc/html/Rsamtools.html</a>.</cite>
</p>
<p>
<a id='bib-GenomicFiles'></a><a href="#cite-GenomicFiles">[19]</a><cite> V. Obenchain, M. Love and M. Morgan. <EM>GenomicFiles: Distributed computing by file or by range</EM>. R package version 1.1.19. 2014.</cite>
</p>
<p>
<a id='bib-AnnotationDbi'></a><a href="#cite-AnnotationDbi">[20]</a><cite> H. Pages, M. Carlson, S. Falcon and N. Li. <EM>AnnotationDbi: Annotation Database Interface</EM>. R package version 1.27.16. 2014.</cite>
</p>
<p>
<a id='bib-S4Vectors'></a><a href="#cite-S4Vectors">[21]</a><cite> H. Pages, M. Lawrence and P. Aboyoun. <EM>S4Vectors: S4 implementation of vectors and lists</EM>. R package version 0.2.4. 2014.</cite>
</p>
<p>
<a id='bib-R'></a><a href="#cite-R">[22]</a><cite> R Core Team. <EM>R: A Language and Environment for Statistical Computing</EM>. R Foundation for Statistical Computing. Vienna, Austria, 2014. URL: <a href="http://www.R-project.org/">http://www.R-project.org/</a>.</cite>
</p>
<p>
<a id='bib-ggplot2'></a><a href="#cite-ggplot2">[23]</a><cite> H. Wickham. <EM>ggplot2: elegant graphics for data analysis</EM>. Springer New York, 2009. ISBN: 978-0-387-98140-6. URL: <a href="http://had.co.nz/ggplot2/book">http://had.co.nz/ggplot2/book</a>.</cite>
</p>
<p>
<a id='bib-testthat'></a><a href="#cite-testthat">[24]</a><cite> H. Wickham. testthat: Get Started with Testing. In: <EM>The R Journal</EM> 3 (2011), pp. 510. URL: <a href="http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>.</cite>
</p>
<p>
<a id='bib-devtools'></a><a href="#cite-devtools">[25]</a><cite> H. Wickham and W. Chang. <EM>devtools: Tools to make developing R code easier</EM>. R package version 1.6. 2014. URL: <a href="http://CRAN.R-project.org/package=devtools">http://CRAN.R-project.org/package=devtools</a>.</cite>
</p>
<p>
<a id='bib-knitr'></a><a href="#cite-knitr">[26]</a><cite> Y. Xie. knitr: A Comprehensive Tool for Reproducible Research in R. In: <EM>Implementing Reproducible Computational Research</EM>. Ed. by V. Stodden, F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman and Hall/CRC, 2014. URL: <a href="http://www.crcpress.com/product/isbn/9781466561595">http://www.crcpress.com/product/isbn/9781466561595</a>.</cite>
</p>
<p>
<a id='bib-biovizBase'></a><a href="#cite-biovizBase">[27]</a><cite> T. Yin, M. Lawrence and D. Cook. <EM>biovizBase: Basic graphic utilities for visualization of genomic data.</EM> R package version 1.13.11. 2014.</cite>
</p>
</div>
</div>
</div>
<div class="navbar navbar-fixed-bottom navbar-inverse">
<div class="container">
<div class="navbar-header">
<button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-responsive-collapse">
<span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span>
</button>
</div>
<div id="bottom-navbar" class="navbar-collapse collapse navbar-responsive-collapse">
<ul class="nav navbar-nav navbar-right">
<li class="nav">
<p class="navbar-text">
Toggle
</p>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Code <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Languages
</li>
<li class="active">
<a href="#" class="toggle-global source R" type="source.R">R</a>
</li>
<li>
<a href="#" type="all-source" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Output <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Type
</li>
<li class="active">
<a href="#" class="toggle-global message" type="message">message</a>
</li>
<li class="active">
<a href="#" class="toggle-global output" type="output">output</a>
</li>
<li>
<a href="#" type="all-output" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="active">
<a href="#" type="figure" class="toggle-global">Figures</a>
</li>
</ul>
<ul class="nav navbar-nav navbar-left">
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Bootstrap theme <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="active">
<a class="theme-switch" title="Default" href="#">Default</a>
</li>
<li>
<a class="theme-switch" title="Amelia" href="#">Amelia</a>
</li>
<li>
<a class="theme-switch" title="Cerulean" href="#">Cerulean</a>
</li>
<li>
<a class="theme-switch" title="Cosmo" href="#">Cosmo</a>
</li>
<li>
<a class="theme-switch" title="Cyborg" href="#">Cyborg</a>
</li>
<li>
<a class="theme-switch" title="Journal" href="#">Journal</a>
</li>
<li>
<a class="theme-switch" title="Flatly" href="#">Flatly</a>
</li>
<li>
<a class="theme-switch" title="Readable" href="#">Readable</a>
</li>
<li>
<a class="theme-switch" title="Simplex" href="#">Simplex</a>
</li>
<li>
<a class="theme-switch" title="Slate" href="#">Slate</a>
</li>
<li>
<a class="theme-switch" title="Spacelab" href="#">Spacelab</a>
</li>
<li>
<a class="theme-switch" title="United" href="#">United</a>
</li>
</ul>
</li>
</ul>
<ul class="nav navbar-nav navbar-left">
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Code highlighting <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="active">
<a class="highlight-switch" title="HighlightJs" href="#">HighlightJs</a>
</li>
<li>
<a class="highlight-switch" title="Dark" href="#">Dark</a>
</li>
<li>
<a class="highlight-switch" title="Far" href="#">Far</a>
</li>
<li>
<a class="highlight-switch" title="Idea" href="#">Idea</a>
</li>
<li>
<a class="highlight-switch" title="Sunburst" href="#">Sunburst</a>
</li>
<li>
<a class="highlight-switch" title="Zenburn" href="#">Zenburn</a>
</li>
<li>
<a class="highlight-switch" title="Visual Studio" href="#">Visual Studio</a>
</li>
<li>
<a class="highlight-switch" title="Ascetic" href="#">Ascetic</a>
</li>
<li>
<a class="highlight-switch" title="Magula" href="#">Magula</a>
</li>
<li>
<a class="highlight-switch" title="Github" href="#">Github</a>
</li>
<li>
<a class="highlight-switch" title="Google Code" href="#">Google Code</a>
</li>
<li>
<a class="highlight-switch" title="Brown Paper" href="#">Brown Paper</a>
</li>
<li>
<a class="highlight-switch" title="School Book" href="#">School Book</a>
</li>
<li>
<a class="highlight-switch" title="IR Black" href="#">IR Black</a>
</li>
<li>
<a class="highlight-switch" title="Solarized - Dark" href="#">Solarized - Dark</a>
</li>
<li>
<a class="highlight-switch" title="Solarized - Light" href="#">Solarized - Light</a>
</li>
<li>
<a class="highlight-switch" title="Arta" href="#">Arta</a>
</li>
<li>
<a class="highlight-switch" title="Monokai" href="#">Monokai</a>
</li>
<li>
<a class="highlight-switch" title="Xcode" href="#">Xcode</a>
</li>
<li>
<a class="highlight-switch" title="Pojoaque" href="#">Pojoaque</a>
</li>
<li>
<a class="highlight-switch" title="Rainbow" href="#">Rainbow</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow" href="#">Tomorrow</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night" href="#">Tomorrow Night</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Bright" href="#">Tomorrow Night Bright</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Blue" href="#">Tomorrow Night Blue</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Eighties" href="#">Tomorrow Night Eighties</a>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" title="Default" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/amelia/bootstrap.min.css" title="Amelia" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cerulean/bootstrap.min.css" title="Cerulean" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cosmo/bootstrap.min.css" title="Cosmo" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cyborg/bootstrap.min.css" title="Cyborg" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/journal/bootstrap.min.css" title="Journal" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/flatly/bootstrap.min.css" title="Flatly" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/readable/bootstrap.min.css" title="Readable" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/simplex/bootstrap.min.css" title="Simplex" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/slate/bootstrap.min.css" title="Slate" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/spacelab/bootstrap.min.css" title="Spacelab" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/united/bootstrap.min.css" title="United" media="screen"></link><link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" title="HighlightJs" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/dark.min.css" title="Dark" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/far.min.css" title="Far" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/idea.min.css" title="Idea" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/sunburst.min.css" title="Sunburst" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/zenburn.min.css" title="Zenburn" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/vs.min.css" title="Visual Studio" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/ascetic.min.css" title="Ascetic" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/magula.min.css" title="Magula" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/github.min.css" title="Github" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/googlecode.min.css" title="Google Code" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/brown_paper.min.css" title="Brown Paper" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/school_book.min.css" title="School Book" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/ir_black.min.css" title="IR Black" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/solarized_dark.min.css" title="Solarized - Dark" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/solarized_light.min.css" title="Solarized - Light" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/arta.min.css" title="Arta" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/monokai.min.css" title="Monokai" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/xcode.min.css" title="Xcode" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/pojoaque.min.css" title="Pojoaque" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/rainbow.min.css" title="Rainbow" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow.min.css" title="Tomorrow" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night.min.css" title="Tomorrow Night" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-bright.min.css" title="Tomorrow Night Bright" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-blue.min.css" title="Tomorrow Night Blue" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-eighties.min.css" title="Tomorrow Night Eighties" media="screen"></link>
</div>
<div id="push">

</div>
<div id="footer">
<div class="container">
<p class="text-muted" id="credit">
Styled with <a href="https://github.com/jimhester/knitrBootstrap">knitrBootstrap</a>
</p>
</div>
</div>
<link rel="stylesheet" id="theme" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" media="screen"></link><link rel="stylesheet" id="highlight" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" media="screen"></link>
</div>
</body>
</html>
